<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="cloneRate">
<title>Analysis of human blood data • cloneRate</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of human blood data">
<meta property="og:description" content="cloneRate">
<meta property="og:image" content="https://bdj34.github.io/cloneRate/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">cloneRate</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/cloneRate-dataAnalysis.html">Analysis of human blood data</a>
    <a class="dropdown-item" href="../articles/cloneRate-simulate.html">Validating growth rate estimates via simulation</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bdj34/cloneRate/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Analysis of human blood data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bdj34/cloneRate/blob/HEAD/vignettes/cloneRate-dataAnalysis.Rmd" class="external-link"><code>vignettes/cloneRate-dataAnalysis.Rmd</code></a></small>
      <div class="d-none name"><code>cloneRate-dataAnalysis.Rmd</code></div>
    </div>

    
    
<p>In this vignette, we’ll walk through our data analysis corresponding
to <a href="https://www.biorxiv.org/" class="external-link">our recent preprint</a> analyzing
single-cell Whole Genome Sequencing (scWGS) data from four different
papers. As part of this package, we have included everything needed to
reproduce this analysis, which consists of reconstructed phylogenetic
trees and de-identified metadata about the donors who were sequenced. We
also include some longitudinal bulk Whole Genome Sequencing (bulk WGS)
data which we use for validation.</p>
<div class="section level3">
<h3 class="unnumbered" id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h3>
<p>First, we’ll have to load the packages we want to use. We’ll be
plotting the trees using the <a href="https://rdrr.io/cran/ape/" class="external-link"><code>ape</code></a> package function
<a href="https://rdrr.io/cran/ape/man/plot.phylo.html" class="external-link"><code>ape::plot.phylo()</code></a>
along with some other <code>ape</code> functions. If you have
<code>cloneRate</code> installed, you already have the <a href="https://rdrr.io/cran/ape/" class="external-link"><code>ape package</code></a>. We’ll
also be using <a href="https://ggplot2.tidyverse.org/" class="external-link"><code>ggplot2</code></a> to make
our plots, as well as <a href="https://rdrr.io/cran/survival/" class="external-link"><code>survival</code></a>, <a href="https://rdrr.io/cran/ggsurvfit/src/R/ggsurvfit.R" class="external-link"><code>ggsurvfit</code></a>,
and <a href="https://rdrr.io/cran/car/" class="external-link"><code>car</code></a> for some
downstream analysis. All of these can be installed from CRAN as shown
below:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load and attach our package cloneRate</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bdj34/cloneRate" class="external-link">cloneRate</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load and attach ape</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://ape-package.ird.fr/" class="external-link">ape</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install ggplot2 if necessary, then load and attach it with library()</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install survival if necessary, then load and attach it with library()</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"survival"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"survival"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/therneau/survival" class="external-link">survival</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install ggsurvfit if necessary, then load and attach it with library()</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"ggsurvfit"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ggsurvfit"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ddsjoberg/ggsurvfit" class="external-link">ggsurvfit</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install car if necessary, then load and attach it with library()</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"car"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"car"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-forge.r-project.org/projects/car/" class="external-link">car</a></span><span class="op">)</span></span></code></pre></div>
<p>We’ll also set the colors which we’ll use for most of the plotting.
The palette is taken from <a href="https://davidmathlogic.com/colorblind/#%23000000-%23E69F00-%2356B4E9-%23009E73-%23F0E442-%230072B2-%23D55E00-%23CC79A7" class="external-link">here</a></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">colorPal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#000000"</span>, <span class="st">"#E69F00"</span>, <span class="st">"#56B4E9"</span>, <span class="st">"#009E73"</span>, <span class="st">"#F0E442"</span>, <span class="st">"#0072B2"</span>, <span class="st">"#D55E00"</span>, <span class="st">"#CC79A7"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data-exploration">Data exploration<a class="anchor" aria-label="anchor" href="#data-exploration"></a>
</h2>
<p>First, let’s examine <code>realCloneData</code>, which contains most
of the necessary data to reproduce our analysis.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu">cloneRate</span><span class="fu">::</span><span class="va"><a href="../reference/realCloneData.html">realCloneData</a></span><span class="op">)</span></span>
<span><span class="co">#&gt;            Length Class  Mode</span></span>
<span><span class="co">#&gt; fullTrees  30     -none- list</span></span>
<span><span class="co">#&gt; cloneTrees 42     -none- list</span></span></code></pre></div>
<p>There are two lists contained in this dataset, one which contains the
full reconstructed tree for each individual at each timepoint (named
<code>fullTrees</code>), and one which contains specific clones from the
full trees (named <code>cloneTrees</code>). We require that a clone must
have at least <code>n=10</code> tips (cells) to be included in
<code>cloneTrees</code>. We discuss this requirement in our <a href="cloneRate-simulate.html">simulation vignette</a>. For now, let’s
split these into two different variables for clarity.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fullTrees.list</span> <span class="op">&lt;-</span> <span class="fu">cloneRate</span><span class="fu">::</span><span class="va"><a href="../reference/realCloneData.html">realCloneData</a></span><span class="op">$</span><span class="va">fullTrees</span></span>
<span><span class="va">cloneTrees.list</span> <span class="op">&lt;-</span> <span class="fu">cloneRate</span><span class="fu">::</span><span class="va"><a href="../reference/realCloneData.html">realCloneData</a></span><span class="op">$</span><span class="va">cloneTrees</span></span></code></pre></div>
<div class="section level3">
<h3 id="full-trees">Full trees<a class="anchor" aria-label="anchor" href="#full-trees"></a>
</h3>
<p>Let’s start by looking at the full tree from a single individual,
“PD9478”. Because some individuals in this study are sequenced at
multiple timepoints, we refer to the ID (“PD9478”) and timepoint (“_1”)
to get a specific tree from <code>fullTrees.list</code>. In this case,
the only timepoint is timepoint 1, so let’s take a look at the full tree
“PD9478_1”. This data comes from the work of <a href="https://pubmed.ncbi.nlm.nih.gov/35058638/" class="external-link">Williams et al.</a></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Let's see the class of the tree</span></span>
<span><span class="va">PD9478</span> <span class="op">&lt;-</span> <span class="va">fullTrees.list</span><span class="op">$</span><span class="va">PD9478_1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">PD9478</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "phylo"</span></span></code></pre></div>
<p>We see that the tree is of class <code>phylo</code>(<a href="http://ape-package.ird.fr/misc/FormatTreeR_24Oct2012.pdf" class="external-link uri">http://ape-package.ird.fr/misc/FormatTreeR_24Oct2012.pdf</a>),
which is a fairly straightforward encoding of phylogenetic trees in
R.</p>
<p>Let’s plot the tree:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot the full tree from individual PD9478 at timepoint 1</span></span>
<span><span class="va">PD9478</span> <span class="op">&lt;-</span> <span class="va">fullTrees.list</span><span class="op">$</span><span class="va">PD9478_1</span></span>
<span></span>
<span><span class="co"># Plot, then add scale and title</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/plot.phylo.html" class="external-link">plot.phylo</a></span><span class="op">(</span><span class="va">PD9478</span>, direction <span class="op">=</span> <span class="st">"downwards"</span>, </span>
<span>            show.tip.label <span class="op">=</span> <span class="cn">FALSE</span>, edge.width <span class="op">=</span> <span class="fl">2</span>,</span>
<span>            edge.color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="fl">2</span><span class="op">)</span>, <span class="st">"red"</span>, <span class="st">"#0070C0"</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="op">(</span><span class="fl">200</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/axisPhylo.html" class="external-link">axisPhylo</a></span><span class="op">(</span>side <span class="op">=</span> <span class="fl">2</span>, backward <span class="op">=</span> <span class="cn">FALSE</span>, las <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/title.html" class="external-link">title</a></span><span class="op">(</span>main<span class="op">=</span><span class="st">"PD9478 full tree"</span>, ylab<span class="op">=</span><span class="st">"Time (years)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="cloneRate-dataAnalysis_files/figure-html/firstTreePlot-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="clone-trees">Clone trees<a class="anchor" aria-label="anchor" href="#clone-trees"></a>
</h3>
<p>If you went through the <a href="cloneRate-simulate.html">simulation
vignette</a>, you’ll notice that this tree looks quite a bit different
than the ones we simulated. In the above tree, there is a clear example
of a somatic clonal expansion. We know from the authors who produced
this data (see <a href="https://www.nature.com/articles/s41586-021-04312-6/figures/3" class="external-link">Williams
et al. Fig. 3 PD9478</a>), that all the sampled cells within this clone
have a mutation in the JAK2 gene (which occurs along the red edge) and
the DNMT3A gene (which occurs along the blue edge). Both of these
mutations are known to be important in clonal expansions in blood. Let’s
look more closely at the clone and see what we can learn from the
tree.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the clone tree from our cloneTrees.list</span></span>
<span><span class="va">PD9478_subClone</span> <span class="op">&lt;-</span> <span class="va">cloneTrees.list</span><span class="op">[[</span><span class="st">"PD9478_1_clone1"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Plot, then add scale and title</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/plot.phylo.html" class="external-link">plot.phylo</a></span><span class="op">(</span><span class="va">PD9478_subClone</span>, direction <span class="op">=</span> <span class="st">"downwards"</span>, </span>
<span>                show.tip.label <span class="op">=</span> <span class="cn">FALSE</span>, edge.width <span class="op">=</span> <span class="fl">2</span>,<span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/axisPhylo.html" class="external-link">axisPhylo</a></span><span class="op">(</span>side <span class="op">=</span> <span class="fl">2</span>, backward <span class="op">=</span> <span class="cn">FALSE</span>, las <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/title.html" class="external-link">title</a></span><span class="op">(</span>main<span class="op">=</span><span class="st">"PD9478 JAK2 &amp; DNMT3A clone tree"</span>, ylab<span class="op">=</span><span class="st">"Time (years)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="cloneRate-dataAnalysis_files/figure-html/plotSubclone-1.png" width="100%"></p>
<p>Now this tree of the clonal expansion looks a lot more like the trees
we simulate! This suggests that the growth rate of each cell within this
clone is roughly the same as other cells within the clone. We can now
apply our methods to the clone tree shown above!</p>
</div>
</div>
<div class="section level2">
<h2 id="estimating-growth-rates">Estimating growth rates<a class="anchor" aria-label="anchor" href="#estimating-growth-rates"></a>
</h2>
<p>We see that the above tree is ultrametric and we know that it’s
time-based, in units of years, so we should apply our
[internalLengths()] and [maxLikelihood()] functions.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get maximum likelihood and internal lengths estimates</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="../reference/maxLikelihood.html">maxLikelihood</a></span><span class="op">(</span><span class="va">PD9478_subClone</span><span class="op">)</span><span class="op">$</span><span class="va">estimate</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.516</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="../reference/internalLengths.html">internalLengths</a></span><span class="op">(</span><span class="va">PD9478_subClone</span><span class="op">)</span><span class="op">$</span><span class="va">estimate</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.601</span></span></code></pre></div>
<p>Unfortunately, we don’t have a ground truth to compare to when
working with real data. However, this individual clone has longitudinal
sequencing data, which we can use to estimate an orthogonal growth rate.
We’ll come back to this later (see <a href="#validation">Validation</a>).</p>
<p>First, let’s apply our analysis to all of the clone trees. If we give
a list of <a href="http://ape-package.ird.fr/misc/FormatTreeR_24Oct2012.pdf" class="external-link"><code>phylo</code></a>
objects, our growth rate functions will estimate the growth rate of all
of them separately, returning a data.frame with each row telling us the
result from each tree.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Apply each of our estimates</span></span>
<span><span class="va">resultsLengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/internalLengths.html">internalLengths</a></span><span class="op">(</span><span class="va">cloneTrees.list</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in FUN(X[[i]], ...): Tree is not binary. Birth-death branching trees should be binary,</span></span>
<span><span class="co">#&gt;        but tree resonstruction from data may lead to  3+ descendants from</span></span>
<span><span class="co">#&gt;        a single parent node. Proceed with caution! Input tree has</span></span>
<span><span class="co">#&gt;       3 nodes directly descending from a single</span></span>
<span><span class="co">#&gt;        parent node. A binary tree would only have 2 descendant nodes</span></span>
<span><span class="co">#&gt;        from each parent node. Tree throwing warning is PD5847_1_clone1</span></span>
<span><span class="co">#&gt; Warning in FUN(X[[i]], ...): Tree is not binary. Birth-death branching trees should be binary,</span></span>
<span><span class="co">#&gt;        but tree resonstruction from data may lead to  3+ descendants from</span></span>
<span><span class="co">#&gt;        a single parent node. Proceed with caution! Input tree has</span></span>
<span><span class="co">#&gt;       3 nodes directly descending from a single</span></span>
<span><span class="co">#&gt;        parent node. A binary tree would only have 2 descendant nodes</span></span>
<span><span class="co">#&gt;        from each parent node. Tree throwing warning is PD4781_1_clone1</span></span></code></pre></div>
<p>Oof! Looks like we have a warning…</p>
<p>It says that two of the trees we input “PD5847_1_clone1” and
“PD4781_1_clone1” are not binary branching trees. This means that one
node has more than 2 direct descendants. In this case, it’s not a big
deal. We know that these trees are resconstructed from mutation data, so
this could just mean that we picked up two divisions which happened
without mutations occurring in between. Let’s plot one of them so we can
see where this happens</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot, highlighting the 3 descendants of node 76. Then add scale and title</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/plot.phylo.html" class="external-link">plot.phylo</a></span><span class="op">(</span><span class="va">cloneTrees.list</span><span class="op">$</span><span class="va">PD5847_1_clone1</span>, edge.width <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                direction <span class="op">=</span> <span class="st">"downwards"</span>, show.tip.label <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                edge.color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span><span class="op">)</span>, <span class="st">"darkorange"</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">34</span><span class="op">)</span>, </span>
<span>                               <span class="st">"darkorange"</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">6</span><span class="op">)</span>, <span class="st">"darkorange"</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1000</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/axisPhylo.html" class="external-link">axisPhylo</a></span><span class="op">(</span>side <span class="op">=</span> <span class="fl">2</span>, backward <span class="op">=</span> <span class="cn">FALSE</span>, las <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/title.html" class="external-link">title</a></span><span class="op">(</span>main<span class="op">=</span><span class="st">"PD5847 clone 1 tree (multi-branching)"</span>, ylab<span class="op">=</span><span class="st">"Time (years)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="cloneRate-dataAnalysis_files/figure-html/plotNonBinary-1.png" width="100%"></p>
<p>We can see the three descendants from a single node (node 76), which
is identified by the edges in orange. We see that there are a lot of
places where there are multiple splits (coalescence events) in quick
succession, but this is the only place where 1 node appears to split
into 3. Well, it’s okay, we know mutations are an imperfect way to keep
track of time. If we saw more of these events where one node splits into
3, or cases where one node splits into 4+, we might be more concerned.
In this case, we can move on. Our functions for growth rate estimation
are equipped to handle trees like this, but they throw warnings so we
know what’s happening. Let’s continue, applying our [maxLikelihood()]
function to the trees.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resultsMaxLike</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="../reference/maxLikelihood.html">maxLikelihood</a></span><span class="op">(</span><span class="va">cloneTrees.list</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Preview the output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">resultsMaxLike</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;                lowerBound  estimate upperBound</span></span>
<span><span class="co">#&gt; PD34493_clone1 0.09732233 0.1687267  0.2401312</span></span>
<span><span class="co">#&gt; PD34493_clone3 0.17680977 0.2316003  0.2863907</span></span>
<span><span class="co">#&gt; PD34493_clone4 0.09428111 0.1957285  0.2971758</span></span>
<span><span class="co">#&gt; PD41305_clone1 0.17392864 0.3301275  0.4863263</span></span>
<span><span class="co">#&gt; PD41305_clone2 0.15906481 0.2757690  0.3924733</span></span>
<span><span class="co">#&gt; PD41305_clone3 0.32741598 0.5434498  0.7594836</span></span>
<span></span>
<span><span class="co"># Print correlation coefficient of the estimates from the two methods</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.test.html" class="external-link">cor.test</a></span><span class="op">(</span><span class="va">resultsLengths</span><span class="op">$</span><span class="va">estimate</span>, <span class="va">resultsMaxLike</span><span class="op">$</span><span class="va">estimate</span><span class="op">)</span><span class="op">$</span><span class="va">estimate</span><span class="op">)</span></span>
<span><span class="co">#&gt;       cor </span></span>
<span><span class="co">#&gt; 0.9829619</span></span></code></pre></div>
<p>Not bad! With a correlation coefficient around 0.98, the estimates
from each method agree very well.</p>
<div class="section level3">
<h3 id="understanding-the-function-outputs">Understanding the function outputs<a class="anchor" aria-label="anchor" href="#understanding-the-function-outputs"></a>
</h3>
<p>Each of the functions ([maxLikelihood()] and [internalLengths()])
returns a data.frame with the same columns, and we’ll walk through what
the columns mean here:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check that column names are the same</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">resultsMaxLike</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">resultsLengths</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Print column names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">resultsMaxLike</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "lowerBound"         "estimate"           "upperBound"        </span></span>
<span><span class="co">#&gt;  [4] "cloneAgeEstimate"   "sumInternalLengths" "sumExternalLengths"</span></span>
<span><span class="co">#&gt;  [7] "extIntRatio"        "n"                  "alpha"             </span></span>
<span><span class="co">#&gt; [10] "runtime_s"          "method"             "cloneName_result"</span></span></code></pre></div>
<ul>
<li>
<code>lowerBound</code>, <code>estimate</code>,
<code>upperBound</code>: These are the growth rate estimate and
confidence intervals. The default is 95% confidence intervals, but this
can be adjusted by setting the <code>alpha</code> parameter in the
function input.</li>
<li>
<code>cloneAgeEstimate</code> is the estimate of the time from clone
initiation to sampling. The earliest coalescence time in the clone tree
gives us a rough estimate of the clone age, and then we add <span class="math inline">\(1/r\)</span>,which is 1 divided by the growth rate
estimate, to account for the expected time from initiation to the first
surviving division (first coalescence). We’ll discuss this more
later.</li>
<li>
<code>sumInternalLengths</code>, <code>sumExternalLengths</code>,
<code>extintRatio</code>: These are the sum of the internal edge
lengths, external edge lengths, and the ratio between the two. We use
this ratio to say whether enough time has passed to use our method (<a href="cloneRate-simulate.html">see simulation vignette</a>).
Fortunately, all of our real data has a ratio &gt; 3, so we can apply
our methods to all of the clone trees.</li>
<li>
<code>n</code> is the number of tips in the clone tree. In this
case, this corresponds to the number of cells sequenced from the
clone.</li>
<li>
<code>alpha</code> is the confidence level. The default is 0.05, and
we use <code>1-alpha</code> (default 95%) confidence intervals for the
upper and lower bounds.</li>
<li>
<code>runtime_s</code> is the runtime of the method in seconds.</li>
<li>
<code>method</code> is the method used, which helps if we want to
combine data.frames from different methods for plotting.</li>
<li>
<code>cloneName_result</code> is a unique name for the clone
corresponding to the individual ID and clone number. For individuals
from the <a href="https://pubmed.ncbi.nlm.nih.gov/35058638/" class="external-link">Williams
paper</a>, where some individuals were sampled at multiple timepoints,
the ID will be followed by “_1”, “_2”, or “_3” to indicate the
timepoint.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="downstream-analysis">Downstream analysis<a class="anchor" aria-label="anchor" href="#downstream-analysis"></a>
</h2>
<div class="section level3">
<h3 id="metadata">Metadata<a class="anchor" aria-label="anchor" href="#metadata"></a>
</h3>
<p>Now that we have growth rates, let’s make some sense of them. If we
look inside cloneTrees.list, we realize that we also have metadata.
Let’s print two examples, one from an individual with a
Myeloproliferative Neoplasm (MPN), and one without:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Show the metadata of an individual without MPN</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">cloneTrees.list</span><span class="op">$</span><span class="va">PD34493_clone1</span><span class="op">$</span><span class="va">metadata</span><span class="op">)</span></span>
<span><span class="co">#&gt;        ID age cloneName_meta      cloneDriver diagnosis.age malnorm sourcePaper</span></span>
<span><span class="co">#&gt; 1 PD34493  83 PD34493_clone1 SF3B1:k666n_DelY            NA  Normal       fabre</span></span>
<span></span>
<span><span class="co"># And the metadata of an individual with MPN</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">cloneTrees.list</span><span class="op">$</span><span class="va">PD9478_1_clone1</span><span class="op">$</span><span class="va">metadata</span><span class="op">)</span></span>
<span><span class="co">#&gt;         ID      age  cloneName_meta                                cloneDriver</span></span>
<span><span class="co">#&gt; 1 PD9478_1 68.75291 PD9478_1_clone1 JAK2:p.F537_K539delinsL_AND_DNMT3A:p.Y908*</span></span>
<span><span class="co">#&gt;   diagnosis.age   malnorm sourcePaper</span></span>
<span><span class="co">#&gt; 1            53 Malignant    williams</span></span></code></pre></div>
<p>For a biological and/or clinical analysis, the most important of
these are:</p>
<ul>
<li>
<code>age</code> which tells us the individual’s age at the time of
sampling. We know that clonal hematopoiesis is associated with
aging.</li>
<li>
<code>cloneDriver</code> which tells us the clonal hematopoiesis
driver which likely leads to the expansion.</li>
<li>
<code>diagnosis.age</code> which tells us when the individual was
diagnosed with Myeloproliferative Neoplasm (MPN). If the individual
doesn’t have MPN, <code>diagnosis.age</code> is <code>NA</code>.</li>
<li>
<code>malnorm</code> which tells us whether the individual has been
diagnosed with a Myeloproliferative Neoplasm (MPN). This information is
already conveyed in <code>diagnosis.age</code>, but it’s sometimes
easier to work with in this format.</li>
</ul>
<p>Let’s combine all of the <code>cloneTrees.list</code> metadata into
one data.frame, so we can analyze it together with the results of
running our functions.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Combine all metadata into a single data.frame</span></span>
<span><span class="va">metadataAll</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">cloneTrees.list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">metadata</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Let’s join the data.frames containing the metadata and the results.
We can use <code><a href="https://rdrr.io/r/base/cbind.html" class="external-link">base::cbind()</a></code> because we know the order of the
metadata is the same as the order of the results data.frame. We can
check this by comparing the <code>cloneName_result</code> and
<code>cloneName_meta</code> columns, which should be (and are) the
same.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Combine metadata with estimates using cbind. Check if cloneNames match</span></span>
<span><span class="va">resultsLengthsMeta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">resultsLengths</span>, <span class="va">metadataAll</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="va">resultsLengthsMeta</span><span class="op">$</span><span class="va">cloneName_result</span> <span class="op">==</span> <span class="va">resultsLengthsMeta</span><span class="op">$</span><span class="va">cloneName_meta</span><span class="op">)</span></span>
<span></span>
<span><span class="va">resultsMaxLikeMeta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">resultsMaxLike</span>, <span class="va">metadataAll</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="va">resultsLengthsMeta</span><span class="op">$</span><span class="va">cloneName_result</span> <span class="op">==</span> <span class="va">resultsLengthsMeta</span><span class="op">$</span><span class="va">cloneName_meta</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Because max. likelihood performs slightly better, use that going forward</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="va">resultsMaxLikeMeta</span></span></code></pre></div>
<p>We’ll go forward with our analysis using the results data.frame,
which contains the estimates from the [maxLikelihood()] function and the
metadata. In the <a href="cloneRate-simulate.html">simulation
vignette</a>, we show that the maximum likelihood estimate performs the
best.</p>
</div>
<div class="section level3">
<h3 id="recreating-figures">Recreating figures<a class="anchor" aria-label="anchor" href="#recreating-figures"></a>
</h3>
<p>Here, we recreate the analysis in <a href="https://www.biorxiv.org/" class="external-link">Figure 5E-G of our work</a>. These are
just a few of many analyses that can be done with this data, so don’t be
limited by what we do.</p>
<p>If we split the cohort by MPN and normal, we might expect differences
in the growth rates. Because the number of sampled cells, on average, is
higher in the healthy individuals due to study design differences, we’re
likely to pick up more clones from the healthy individuals. To avoid
this bias as much as possible, we’ll compare the mean growth rate
between individuals with and without MPN.</p>
<p>First, we want to avoid double counting the clones which appear
twice. There are three such clones from the <a href="https://pubmed.ncbi.nlm.nih.gov/35058638/" class="external-link">Williams paper</a>:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># If patient ID and clone number are the same, even if timepoint </span></span>
<span><span class="co"># differs, then we have a duplicate.</span></span>
<span><span class="co"># Make a new column for patient, removing anything after "_" from cloneName_result</span></span>
<span><span class="va">results</span><span class="op">$</span><span class="va">patient</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_.*"</span>, <span class="st">""</span>, <span class="va">results</span><span class="op">$</span><span class="va">cloneName_result</span><span class="op">)</span></span>
<span><span class="va">results</span><span class="op">$</span><span class="va">cloneNumber</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">".*_clone"</span>, <span class="st">""</span>, <span class="va">results</span><span class="op">$</span><span class="va">cloneName_result</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combining patient ID (without timepoint) and clone number </span></span>
<span><span class="co"># will give us a unique ID for the clone regardless of sampling time</span></span>
<span><span class="va">results</span><span class="op">$</span><span class="va">uniqueCloneID</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">patient</span>, <span class="st">"_"</span>, <span class="va">results</span><span class="op">$</span><span class="va">cloneNumber</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Find which clone IDs appear twice</span></span>
<span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">uniqueCloneID</span><span class="op">)</span></span>
<span><span class="va">repeatsVec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">tmp</span><span class="op">)</span><span class="op">[</span><span class="va">tmp</span> <span class="op">==</span> <span class="fl">2</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Remove the duplicate with fewer number of cells sequenced, n</span></span>
<span><span class="va">rowsRemove</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">cloneID</span> <span class="kw">in</span> <span class="va">repeatsVec</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">duplicateRows</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">uniqueCloneID</span> <span class="op">==</span> <span class="va">cloneID</span><span class="op">)</span></span>
<span>  <span class="va">removeIndex</span> <span class="op">&lt;-</span> <span class="va">duplicateRows</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.min</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">n</span><span class="op">[</span><span class="va">duplicateRows</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">rowsRemove</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">rowsRemove</span>, <span class="va">removeIndex</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">uniqueResults</span> <span class="op">&lt;-</span> <span class="va">results</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">rowsRemove</span>, <span class="op">]</span></span>
<span><span class="co"># Check that each unique clone now only appears once (no duplicates)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">uniqueResults</span><span class="op">$</span><span class="va">uniqueCloneID</span><span class="op">)</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>For now, our <code>uniqueResults</code> dataframe contains the
results with only one clone from each duplicated pair. Later on, when we
want to validate our estimates, we’ll make use of the fact that we have
data from different timepoints but the same clone, showing that our
methods make consistent estimates regardless of which sampling time we
use. More on that in the <a href="#validation">Validation</a>
section.</p>
<div class="section level4">
<h4 id="mpn-vs--normal-fig--5e">MPN vs. Normal (Fig. 5E)<a class="anchor" aria-label="anchor" href="#mpn-vs--normal-fig--5e"></a>
</h4>
<p>Now, let’s take the mean for each individual and re-create <a href="https://www.biorxiv.org/" class="external-link">Fig. 5E</a>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the unique individuals, and initialize vectors</span></span>
<span><span class="va">uniqueIndividuals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">uniqueResults</span><span class="op">$</span><span class="va">patient</span><span class="op">)</span></span>
<span><span class="va">individualMeans</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">malNorm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fill vectors with mean growth rate and MPN status of each individual</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">ind</span> <span class="kw">in</span> <span class="va">uniqueIndividuals</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">individualMeans</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">individualMeans</span>, <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">uniqueResults</span><span class="op">$</span><span class="va">estimate</span><span class="op">[</span><span class="va">uniqueResults</span><span class="op">$</span><span class="va">patient</span> <span class="op">==</span> <span class="va">ind</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">malNorm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">malNorm</span>, <span class="va">uniqueResults</span><span class="op">$</span><span class="va">malnorm</span><span class="op">[</span><span class="va">uniqueResults</span><span class="op">$</span><span class="va">patient</span> <span class="op">==</span> <span class="va">ind</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Combine results into a data.frame</span></span>
<span><span class="va">mal_vs_norm.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  <span class="st">"Patient"</span> <span class="op">=</span> <span class="va">uniqueIndividuals</span>,</span>
<span>  <span class="st">"meanEstimate"</span> <span class="op">=</span> <span class="va">individualMeans</span>,</span>
<span>  <span class="st">"malNorm"</span> <span class="op">=</span> <span class="va">malNorm</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run a non-parametric Mann-whitney test to see if they're significantly different</span></span>
<span><span class="va">nonparamTest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/wilcox.test.html" class="external-link">wilcox.test</a></span><span class="op">(</span><span class="va">meanEstimate</span> <span class="op">~</span> <span class="va">malNorm</span>, data <span class="op">=</span> <span class="va">mal_vs_norm.df</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">nonparamTest</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Wilcoxon rank sum exact test</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; data:  meanEstimate by malNorm</span></span>
<span><span class="co">#&gt; W = 64, p-value = 0.02853</span></span>
<span><span class="co">#&gt; alternative hypothesis: true location shift is not equal to 0</span></span>
<span></span>
<span><span class="co"># Set factor ordering for plot and plot using ggplot (Fig. 5E)</span></span>
<span><span class="va">mal_vs_norm.df</span><span class="op">$</span><span class="va">malNorm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">mal_vs_norm.df</span><span class="op">$</span><span class="va">malNorm</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Normal"</span>, <span class="st">"Malignant"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">mal_vs_norm.df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">malNorm</span>, y <span class="op">=</span> <span class="va">meanEstimate</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_label</a></span><span class="op">(</span>label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"p = "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">nonparamTest</span><span class="op">$</span><span class="va">p.value</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>, x <span class="op">=</span> <span class="fl">1.5</span>, y <span class="op">=</span> <span class="fl">1.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.3</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">malNorm</span><span class="op">)</span>, outlier.shape <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">malNorm</span>, y <span class="op">=</span> <span class="va">meanEstimate</span>, color <span class="op">=</span> <span class="va">malNorm</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">.1</span>, size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Normal"</span>, <span class="st">"MPN"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Mean net growth rate (r)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    legend.position <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>    axis.ticks.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="cloneRate-dataAnalysis_files/figure-html/meanInd-1.png" width="50%"></p>
</div>
<div class="section level4">
<h4 id="single-vs--multi-driver-fig--5f">Single vs. multi-driver (Fig. 5F)<a class="anchor" aria-label="anchor" href="#single-vs--multi-driver-fig--5f"></a>
</h4>
<p>Taking our results from unique clones, <code>uniqueResults</code>,
let’s compare growth rates from clones with either an unknown driver or
a single driver, to those with multiple drivers. For this, we’ll notice
a handy convention in the metadata: drivers are separated with “AND”, so
any clone with “AND” in the driver column, has multiple drivers:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set the multi driver variable according to whether multiple drivers are annotated</span></span>
<span><span class="va">uniqueResults</span><span class="op">$</span><span class="va">multiDriver</span> <span class="op">&lt;-</span> <span class="st">"Single or unknown"</span></span>
<span><span class="va">uniqueResults</span><span class="op">$</span><span class="va">multiDriver</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"AND"</span>, <span class="va">uniqueResults</span><span class="op">$</span><span class="va">cloneDriver</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Multiple"</span></span>
<span></span>
<span><span class="co"># Run a non-parametric test to see if the differences are significant</span></span>
<span><span class="va">nonParamTest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/wilcox.test.html" class="external-link">wilcox.test</a></span><span class="op">(</span><span class="va">estimate</span> <span class="op">~</span> <span class="va">multiDriver</span>, data <span class="op">=</span> <span class="va">uniqueResults</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">nonParamTest</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Wilcoxon rank sum exact test</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; data:  estimate by multiDriver</span></span>
<span><span class="co">#&gt; W = 276, p-value = 1.554e-06</span></span>
<span><span class="co">#&gt; alternative hypothesis: true location shift is not equal to 0</span></span>
<span></span>
<span><span class="co"># Set factor levels to control plot order</span></span>
<span><span class="va">uniqueResults</span><span class="op">$</span><span class="va">multiDriver</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">uniqueResults</span><span class="op">$</span><span class="va">multiDriver</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Single or unknown"</span>, <span class="st">"Multiple"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot (Fig. 5F)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">uniqueResults</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">multiDriver</span>, y <span class="op">=</span> <span class="va">estimate</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.3</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">multiDriver</span><span class="op">)</span>, outlier.shape <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">multiDriver</span>, y <span class="op">=</span> <span class="va">estimate</span>, color <span class="op">=</span> <span class="va">multiDriver</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">colorPal</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span>, <span class="va">colorPal</span><span class="op">[</span><span class="fl">8</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_label</a></span><span class="op">(</span>label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"p="</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">nonParamTest</span><span class="op">$</span><span class="va">p.value</span>, <span class="fl">7</span><span class="op">)</span><span class="op">)</span>, x <span class="op">=</span> <span class="fl">1.5</span>, y <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Net growth rate (r)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Driver mutations"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    legend.position <span class="op">=</span> <span class="st">"none"</span>, axis.ticks.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">.5</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="cloneRate-dataAnalysis_files/figure-html/multiDriver-1.png" width="50%"></p>
</div>
<div class="section level4">
<h4 id="time-from-clone-initiation-to-diagnosis-fig--5g">Time from clone initiation to diagnosis (<a href="https://www.biorxiv.org/" class="external-link">Fig. 5G</a>)<a class="anchor" aria-label="anchor" href="#time-from-clone-initiation-to-diagnosis-fig--5g"></a>
</h4>
<p>We discussed before that our results data.frame (the output of
[internalLengths()] or [maxLikelihood()]) came with a clone age
estimate, which was <span class="math inline">\(1/r\)</span> + the time
from sampling back to the first coalescence. If the birth rate of a
clone is <span class="math inline">\(a\)</span> and the death rate is
<span class="math inline">\(b\)</span>, then the net growth rate is
<span class="math inline">\(r=a-b\)</span>. The first coalescence
represents the first surviving division because, if neither of the
offspring of a division die out, they’ll account for, on average, half
the cells in the clone. This means that even if we sample only 10 cells,
we’re very likely to sample at least one descendant from each of the two
offspring, meaning that first surviving division should be the earliest
coalescence, relative to the time of sampling. How do we know when the
first surviving division occurs? Well, we know:</p>
<ul>
<li>the birth rate is <span class="math inline">\(a\)</span>
</li>
<li>the probability of survival of a given individual is <span class="math inline">\((a-b)/a\)</span> or <span class="math inline">\(r/a\)</span>
</li>
<li>This leads to a rate of surviving divisions equal to <span class="math inline">\(a*r/a = r\)</span>. Therefore, the expected
waiting time from clone initiation to first surviving division is <span class="math inline">\(1/r\)</span>.</li>
</ul>
<p>With that out of the way, let’s see if the growth rate tells us
anything about the time from clone initiation to diagnosis. For this, we
want to use the max. growth rate from each patient, as that’s the one
which we’d assume to be most likely to drive MPN initiation. We don’t
understand MPN very well though, so an alternative could be to use the
most recently emerged clone prior to diagnosis. We’ll do both here,
showing that they actually give the same results, probably because any
clone which arises later in life must be fit enough to out-compete other
clones which have already expanded.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the clone IDs of the fittest clone in each patient</span></span>
<span><span class="va">cloneIDs_max</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">uniqueResults</span><span class="op">$</span><span class="va">patient</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">patient.df</span> <span class="op">&lt;-</span> <span class="va">uniqueResults</span><span class="op">[</span><span class="va">uniqueResults</span><span class="op">$</span><span class="va">patient</span> <span class="op">==</span> <span class="va">x</span>, <span class="op">]</span></span>
<span>  <span class="va">cloneID</span> <span class="op">&lt;-</span> <span class="va">patient.df</span><span class="op">$</span><span class="va">cloneName_result</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">patient.df</span><span class="op">$</span><span class="va">estimate</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">cloneID</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get the clone IDs of the youngest clone in each patient</span></span>
<span><span class="va">cloneIDs_youngest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">uniqueResults</span><span class="op">$</span><span class="va">patient</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">patient.df</span> <span class="op">&lt;-</span> <span class="va">uniqueResults</span><span class="op">[</span><span class="va">uniqueResults</span><span class="op">$</span><span class="va">patient</span> <span class="op">==</span> <span class="va">x</span>, <span class="op">]</span></span>
<span>  <span class="va">cloneID</span> <span class="op">&lt;-</span> <span class="va">patient.df</span><span class="op">$</span><span class="va">cloneName_result</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.min</a></span><span class="op">(</span><span class="va">patient.df</span><span class="op">$</span><span class="va">cloneAgeEstimate</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">cloneID</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># See how much overlap there is between the most recent clone and the highest growth rate</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">cloneIDs_max</span> <span class="op">==</span> <span class="va">cloneIDs_youngest</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; FALSE  TRUE </span></span>
<span><span class="co">#&gt;     4    15</span></span></code></pre></div>
<p>We see that in 15/19 cases, the youngest clone (with &gt;= 10 tips)
is also the most fit. First, let’s use the most fit clones and stratify
by the mean, reproducing <a href="https://www.biorxiv.org/" class="external-link">our result
in Figure 5G.</a></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Subset the results to only have the fittest clone for each patient</span></span>
<span><span class="va">maxResults</span> <span class="op">&lt;-</span> <span class="va">uniqueResults</span><span class="op">[</span><span class="va">uniqueResults</span><span class="op">$</span><span class="va">cloneName_result</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">cloneIDs_max</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Subset to the MPN patients</span></span>
<span><span class="va">malMax</span> <span class="op">&lt;-</span> <span class="va">maxResults</span><span class="op">[</span><span class="va">maxResults</span><span class="op">$</span><span class="va">malnorm</span> <span class="op">==</span> <span class="st">"Malignant"</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Determine time from clone initiation to diagnosis</span></span>
<span><span class="va">malMax</span><span class="op">$</span><span class="va">latency_to_dx</span> <span class="op">&lt;-</span> <span class="va">malMax</span><span class="op">$</span><span class="va">cloneAgeEstimate</span> <span class="op">-</span> <span class="op">(</span><span class="va">malMax</span><span class="op">$</span><span class="va">age</span> <span class="op">-</span> <span class="va">malMax</span><span class="op">$</span><span class="va">diagnosis.age</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Stratify the population by the mean growth rate, adding a column "aboveMean"</span></span>
<span><span class="va">malMax</span><span class="op">$</span><span class="va">aboveMean</span> <span class="op">&lt;-</span> <span class="va">malMax</span><span class="op">$</span><span class="va">estimate</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">malMax</span><span class="op">$</span><span class="va">estimate</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Use the survival package to see if differences are significant</span></span>
<span><span class="va">malMax</span><span class="op">$</span><span class="va">status</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">survivalTest</span> <span class="op">&lt;-</span> <span class="fu">survival</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/survdiff.html" class="external-link">survdiff</a></span><span class="op">(</span><span class="fu">survival</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span>time <span class="op">=</span> <span class="va">latency_to_dx</span>, event <span class="op">=</span> <span class="va">status</span><span class="op">)</span> <span class="op">~</span> <span class="va">aboveMean</span>, data <span class="op">=</span> <span class="va">malMax</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">survivalTest</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.002628379</span></span>
<span></span>
<span><span class="co"># Plot survival curves</span></span>
<span><span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/survfit2.html" class="external-link">survfit2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span>time <span class="op">=</span> <span class="va">latency_to_dx</span>, event <span class="op">=</span> <span class="va">status</span><span class="op">)</span> <span class="op">~</span> <span class="va">aboveMean</span>, data <span class="op">=</span> <span class="va">malMax</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/ggsurvfit.html" class="external-link">ggsurvfit</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    y <span class="op">=</span> <span class="st">"Probability diagnosis-free"</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Time from clone initiation (yrs.)"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">colorPal</span><span class="op">[</span><span class="fl">7</span><span class="op">]</span>, <span class="va">colorPal</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_label</a></span><span class="op">(</span>label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"p = "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">survivalTest</span><span class="op">$</span><span class="va">pvalue</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span>, x <span class="op">=</span> <span class="fl">45</span>, y <span class="op">=</span> <span class="fl">.85</span>, color <span class="op">=</span> <span class="st">"black"</span>, fill <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_label</a></span><span class="op">(</span>label <span class="op">=</span> <span class="st">"r &lt; mean"</span>, x <span class="op">=</span> <span class="fl">40</span>, y <span class="op">=</span> <span class="fl">.375</span>, color <span class="op">=</span> <span class="va">colorPal</span><span class="op">[</span><span class="fl">7</span><span class="op">]</span>, fill <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_label</a></span><span class="op">(</span>label <span class="op">=</span> <span class="st">"r &gt; mean"</span>, x <span class="op">=</span> <span class="fl">12</span>, y <span class="op">=</span> <span class="fl">.35</span>, color <span class="op">=</span> <span class="va">colorPal</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>, fill <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span></span></code></pre></div>
<p><img src="cloneRate-dataAnalysis_files/figure-html/maxStratify-1.png" width="50%"></p>
<p>Because we don’t fully understand clonal hematopoiesis and MPN, it’s
reasonable to wonder whether using the fittest clone is the best way to
compare time to diagnosis. As an alternative, let’s compare the youngest
clones, seeing if our growth rate estimates still stratify the time to
diagnosis.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">youngestResults</span> <span class="op">&lt;-</span> <span class="va">uniqueResults</span><span class="op">[</span><span class="va">uniqueResults</span><span class="op">$</span><span class="va">cloneName_result</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">cloneIDs_youngest</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Subset to the MPN patients</span></span>
<span><span class="va">malYoungest</span> <span class="op">&lt;-</span> <span class="va">youngestResults</span><span class="op">[</span><span class="va">youngestResults</span><span class="op">$</span><span class="va">malnorm</span> <span class="op">==</span> <span class="st">"Malignant"</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Check how much overlap there is between youngest clone and fittest clone in the MPN dataset</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">malYoungest</span><span class="op">$</span><span class="va">cloneName_result</span> <span class="op">==</span> <span class="va">malMax</span><span class="op">$</span><span class="va">cloneName_result</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; TRUE </span></span>
<span><span class="co">#&gt;   13</span></span></code></pre></div>
<p>As it turns out, the youngest clone is the fittest clone in all 13
individuals diagnosed with MPN. We saw before that this was the case in
15 of the 19 individuals when including the healthy individuals.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="validation">Validation<a class="anchor" aria-label="anchor" href="#validation"></a>
</h2>
<p>Remember when we said that there were three cases of duplicate
clones? Well, if our estimates are worth anything, they must be
consistent. This means that, in these three clones, we should expect
that the estimates are similar at both timepoints. Let’s check this:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Like before, we can find which clone IDs appear twice</span></span>
<span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">uniqueCloneID</span><span class="op">)</span></span>
<span><span class="va">repeatsVec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">tmp</span><span class="op">)</span><span class="op">[</span><span class="va">tmp</span> <span class="op">==</span> <span class="fl">2</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Subset the results to include only repeated clones</span></span>
<span><span class="va">repeatResults</span> <span class="op">&lt;-</span> <span class="va">results</span><span class="op">[</span><span class="va">results</span><span class="op">$</span><span class="va">uniqueCloneID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">repeatsVec</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Plot the repeated clones</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">repeatResults</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html" class="external-link">geom_pointrange</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">repeatResults</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">cloneName_result</span>, y <span class="op">=</span> <span class="va">estimate</span>, ymin <span class="op">=</span> <span class="va">lowerBound</span>,</span>
<span>    ymax <span class="op">=</span> <span class="va">upperBound</span>, color <span class="op">=</span> <span class="va">uniqueCloneID</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">colorPal</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">6</span>, <span class="fl">7</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Net growth rate estimate (r)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="cloneRate-dataAnalysis_files/figure-html/consistency-1.png" width="50%"></p>
<p>Not too bad! It’s good to be consistent, but it’s better to be
consistent and accurate. For that, we need a completely orthogonal way
to measure clonal growth rates, which comes from longitudinal data. One
of the clones from the <a href="https://pubmed.ncbi.nlm.nih.gov/35058638/" class="external-link">Williams paper</a> and
two of the clones from the <a href="https://pubmed.ncbi.nlm.nih.gov/35650444/" class="external-link">Fabre paper</a> have
matching longitudinal data. We have formatted this nicely and it comes
with the package!</p>
<p>Let’s now fit a logistic curve to the longitudinal data and estimate
the growth rate for one of these clones, from individual PD9478, which
is the tree we showed at the beginning of this vignette. Here, we are
recreating some of the analysis from <a href="https://www.biorxiv.org/" class="external-link">our Figure 6.</a></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get data from PD9478</span></span>
<span><span class="va">PD9478_long</span> <span class="op">&lt;-</span> <span class="va">longitudinalData</span><span class="op">[</span><span class="va">longitudinalData</span><span class="op">$</span><span class="va">Sample.ID</span> <span class="op">==</span> <span class="st">"PD9478"</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Get a rough estimate for starting params</span></span>
<span><span class="va">startingParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/coef.html" class="external-link">coef</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/car/man/logit.html" class="external-link">logit</a></span><span class="op">(</span><span class="va">VAF</span> <span class="op">/</span> <span class="fl">.5</span><span class="op">)</span> <span class="op">~</span> <span class="va">Age</span>, data <span class="op">=</span> <span class="va">PD9478_long</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit to a three parameter logistic curve</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/nls.html" class="external-link">nls</a></span><span class="op">(</span><span class="va">VAF</span> <span class="op">~</span> <span class="va">K</span> <span class="op">/</span> <span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="op">-</span><span class="op">(</span><span class="va">phi2</span> <span class="op">+</span> <span class="va">r</span> <span class="op">*</span> <span class="va">Age</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>K <span class="op">=</span> <span class="fl">0.2</span>, phi2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.000001</span>, <span class="va">startingParams</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>, r <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.00001</span>, <span class="va">startingParams</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  data <span class="op">=</span> <span class="va">PD9478_long</span>, trace <span class="op">=</span> <span class="cn">FALSE</span>, algorithm <span class="op">=</span> <span class="st">"port"</span>, lower <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="op">-</span><span class="fl">500</span>, <span class="fl">0.00001</span><span class="op">)</span>, upper <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set output equal to summary of longitudinal models</span></span>
<span><span class="va">output</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assign fitted params</span></span>
<span><span class="va">r</span> <span class="op">&lt;-</span> <span class="va">output</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="st">"r"</span>, <span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">t_m</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="va">output</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="st">"phi2"</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">/</span> <span class="va">r</span> <span class="co"># Get midpoint time by dividing by -1/r</span></span>
<span><span class="va">K</span> <span class="op">&lt;-</span> <span class="va">output</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="st">"K"</span>, <span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Get CI for growth rate of longitudinal/logistic model</span></span>
<span><span class="va">stdError</span> <span class="op">&lt;-</span> <span class="va">output</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="st">"r"</span>, <span class="st">"Std. Error"</span><span class="op">]</span></span>
<span><span class="va">lb</span> <span class="op">&lt;-</span> <span class="va">r</span> <span class="op">-</span> <span class="fl">1.96</span> <span class="op">*</span> <span class="va">stdError</span></span>
<span><span class="va">ub</span> <span class="op">&lt;-</span> <span class="va">r</span> <span class="op">+</span> <span class="fl">1.96</span> <span class="op">*</span> <span class="va">stdError</span></span>
<span></span>
<span><span class="co"># Prepare a df for plotting the fit line along with the data</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">PD9478_long</span><span class="op">$</span><span class="va">Age</span><span class="op">)</span> <span class="op">-</span> <span class="fl">10</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">PD9478_long</span><span class="op">$</span><span class="va">Age</span><span class="op">)</span> <span class="op">+</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span> <span class="co"># construct a range of x values bounded by the data</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">K</span> <span class="op">/</span> <span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="op">-</span><span class="op">(</span><span class="va">x</span> <span class="op">-</span> <span class="va">t_m</span><span class="op">)</span> <span class="op">*</span> <span class="va">r</span><span class="op">)</span><span class="op">)</span> <span class="co"># curve VAF (3 param model)</span></span>
<span><span class="va">predict.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="st">"x"</span> <span class="op">=</span> <span class="va">x</span>, <span class="st">"y"</span> <span class="op">=</span> <span class="va">y</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set color for longitudinal fit info</span></span>
<span><span class="va">fitColor</span> <span class="op">&lt;-</span> <span class="va">colorPal</span><span class="op">[</span><span class="fl">6</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Plot</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">PD9478_long</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Age</span>, y <span class="op">=</span> <span class="va">VAF</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.01</span>, <span class="fl">0.52</span><span class="op">)</span>, expand <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Person Age (yr)"</span>, y <span class="op">=</span> <span class="st">"Variant Allele Frequency (VAF)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_"</span>, <span class="st">" &amp; "</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">PD9478_long</span><span class="op">$</span><span class="va">Sample.ID</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">" "</span>, <span class="va">PD9478_long</span><span class="op">$</span><span class="va">Gene</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">predict.df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span><span class="op">)</span>, color <span class="op">=</span> <span class="va">fitColor</span>, linewidth <span class="op">=</span> <span class="fl">1</span>, show.legend <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"#808080"</span>, shape <span class="op">=</span> <span class="fl">18</span>, size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="va">t_m</span>, linetype <span class="op">=</span> <span class="st">"dotted"</span>, color <span class="op">=</span> <span class="va">fitColor</span>, linewidth <span class="op">=</span> <span class="fl">.6</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="va">K</span>, linetype <span class="op">=</span> <span class="st">"dotted"</span>, color <span class="op">=</span> <span class="va">fitColor</span>, linewidth <span class="op">=</span> <span class="fl">.6</span><span class="op">)</span></span></code></pre></div>
<p><img src="cloneRate-dataAnalysis_files/figure-html/loadLong-1.png" width="50%"></p>
<p>Now, let’s see how the estimate compares with our maximum likelihood
estimate:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the single cell estimate from PD9478 clone 1</span></span>
<span><span class="va">scPD9478</span> <span class="op">&lt;-</span> <span class="va">results</span><span class="op">[</span><span class="va">results</span><span class="op">$</span><span class="va">cloneName_result</span> <span class="op">==</span> <span class="st">"PD9478_1_clone1"</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Combine orthogonal estimates into one df for plotting</span></span>
<span><span class="va">combine.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  <span class="st">"Clone"</span> <span class="op">=</span> <span class="st">"PD9478_1_clone1"</span>,</span>
<span>  <span class="st">"r"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">r</span>, <span class="va">scPD9478</span><span class="op">$</span><span class="va">estimate</span><span class="op">)</span>,</span>
<span>  <span class="st">"lowerBound"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">lb</span>, <span class="va">scPD9478</span><span class="op">$</span><span class="va">lowerBound</span><span class="op">)</span>,</span>
<span>  <span class="st">"upperBound"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">ub</span>, <span class="va">scPD9478</span><span class="op">$</span><span class="va">upperBound</span><span class="op">)</span>,</span>
<span>  <span class="st">"method"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"longitudinal"</span>, <span class="st">"max. likelihood"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">combine.df</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html" class="external-link">geom_pointrange</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">method</span>, y <span class="op">=</span> <span class="va">r</span>, ymin <span class="op">=</span> <span class="va">lowerBound</span>, ymax <span class="op">=</span> <span class="va">upperBound</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Net growth rate (r)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="cloneRate-dataAnalysis_files/figure-html/validateLong-1.png" width="50%"></p>
<p>All in all, not too bad! For now, we’ll leave the others for you to
explore. This vignette has already been a bit long. But now you should
have the tools to reproduce our analysis relating to the real data
aspect of <a href="https://www.biorxiv.org/" class="external-link">our paper</a>. Hopefully
you were able to do this in minutes and hours, which is much better than
the months it took me!</p>
</div>
<div class="section level2">
<h2 id="references-data-sources">References / Data sources:<a class="anchor" aria-label="anchor" href="#references-data-sources"></a>
</h2>
<p>Any real world data we include is produced and annotated in these
fascinating papers, which have advanced the capabilities of single-cell
DNA sequencing. They also provide excellent analysis in addition to what
we provide here and in our work.</p>
<ul>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/35058638/" class="external-link">Williams et
al. 2022</a></li>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/35650442/" class="external-link">Mitchell et
al. 2022</a></li>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/35650444/" class="external-link">Fabre et
al. 2022</a></li>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/33621486/" class="external-link">Van Egeren et
al. 2021</a></li>
</ul>
<p>The mathematical methods for estimating growth rates build in large
part from the work of Amaury Lambert in:</p>
<ul>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/29704514/" class="external-link">Lambert,
2018</a></li>
</ul>
<p>And here’s a final link to <a href="https://www.biorxiv.org/" class="external-link">our
paper</a> for more of the details of the methods and data analysis.</p>
<p>If you enjoyed the color palette for the plotting, we tried to use
colorblind friendly colors, specifically pulling colors from a palette
designed by <a href="https://www.nature.com/articles/nmeth.1618" class="external-link">Bang
Wong</a> and available <a href="https://davidmathlogic.com/colorblind/#%23000000-%23E69F00-%2356B4E9-%23009E73-%23F0E442-%230072B2-%23D55E00-%23CC79A7" class="external-link">here.</a></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Brian Johnson, Yubo Shuai, Jason Schweinsberg, Kit Curtius.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
