<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="cloneRate">
<title>Reproduce simulation results • cloneRate</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Reproduce simulation results">
<meta property="og:description" content="cloneRate">
<meta property="og:image" content="https://bdj34.github.io/cloneRate/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">cloneRate</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/reproduce-simResults.html">Reproduce simulation results</a>
    <a class="dropdown-item" href="../articles/cloneRate-dataAnalysis.html">Analysis of human blood data</a>
    <a class="dropdown-item" href="../articles/cloneRate-simulate.html">Intro to validating growth rate estimates via simulation</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bdj34/cloneRate/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Reproduce simulation results</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bdj34/cloneRate/blob/HEAD/vignettes/articles/reproduce-simResults.Rmd" class="external-link"><code>vignettes/articles/reproduce-simResults.Rmd</code></a></small>
      <div class="d-none name"><code>reproduce-simResults.Rmd</code></div>
    </div>

    
    
<p>NOTE: This is not included as a vignette with the package due to
long-running scripts. Instead, it will only be available as an article
on <a href="https://bdj34.github.io/cloneRate/">this website</a>. This
is intended to reproduce simulations shown in our recent work <a href="https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btad561/7271182" class="external-link">“cloneRate:
fast estimation of single-cell clonal dynamics using coalescent
theory”</a>. We’ll skip over many of the basics, as those are detailed
in our introductory simulation article/vignette which is available in
<code><a href="../articles/cloneRate-simulate.html">vignette("cloneRate-simulate")</a></code>. The simulation procedure
for generating birth-death trees is a direct implementation of results
by Amaury Lambert in <a href="https://pubmed.ncbi.nlm.nih.gov/29704514/" class="external-link">a recent paper</a>.
These results allowed us to simulate thousands of trees to check our
methods for growth rate estimation from phylogenetic tree
reconstruction, which is detailed in <a href="https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btad561/7271182" class="external-link">our
recent paper.</a> We’ll reproduce this validation here.</p>
<div class="section level3">
<h3 class="unnumbered" id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h3>
<p>First, we’ll have to load the packages we want to use. We’ll be
plotting the trees using the <a href="https://rdrr.io/cran/ape/" class="external-link"><code>ape</code></a> package function
<a href="https://rdrr.io/cran/ape/man/plot.phylo.html" class="external-link"><code>ape::plot.phylo()</code></a>
along with some other <code>ape</code> functions. If you have
<code>cloneRate</code> installed, you already have the <a href="https://rdrr.io/cran/ape/" class="external-link"><code>ape package</code></a>. We’ll
also be using <a href="https://ggplot2.tidyverse.org/" class="external-link"><code>ggplot2</code></a> to make
our plots, which can be installed from CRAN as shown below:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load and attach our package cloneRate</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bdj34/cloneRate" class="external-link">cloneRate</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load and attach ape, which will be installed if you've installed cloneRate</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://ape-package.ird.fr/" class="external-link">ape</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load and attach rstan, which is necessary for running MCMC's (birthDeathMCMC() and Phylofit)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mc-stan.org/rstan/" class="external-link">rstan</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load and attach the parallel package, which is not strictly necessary for running MCMC's, but greatly reduces the time required and is necessary for executing code in this article.</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">parallel</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install ggplot2 if necessary, then load and attach it with library()</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<p>We’ll also set the color palette which we’ll use for most of the
plotting. The palette is taken from <a href="https://davidmathlogic.com/colorblind/#%23000000-%23E69F00-%2356B4E9-%23009E73-%23F0E442-%230072B2-%23D55E00-%23CC79A7" class="external-link">here</a></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">colorPal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#000000"</span>, <span class="st">"#E69F00"</span>, <span class="st">"#56B4E9"</span>, <span class="st">"#009E73"</span>, <span class="st">"#F0E442"</span>, <span class="st">"#0072B2"</span>, <span class="st">"#D55E00"</span>, <span class="st">"#CC79A7"</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 class="unnumbered" id="phylofit-setup">Phylofit setup<a class="anchor" aria-label="anchor" href="#phylofit-setup"></a>
</h4>
<p>For applying Phylofit, which was developed by Nicholas Williams and
applied in <a href="https://www.nature.com/articles/s41586-021-04312-6" class="external-link">Williams et
al.</a> and <a href="https://www.nature.com/articles/s41586-022-04786-y" class="external-link">Mitchell et
al.</a>, see Emily Mitchell’s github: <a href="https://github.com/emily-mitchell/normal_haematopoiesis/blob/main/1_functions/farm_only/phylofit.R" class="external-link uri">https://github.com/emily-mitchell/normal_haematopoiesis/blob/main/1_functions/farm_only/phylofit.R</a>.
For reproducibility, we’ll include the code from Emily Mithcell’s github
in a slightly simplified form, as we don’t want to apply the Aberrant
Cell Fraction (ACF) implementation of Phylofit, for reasons detailed in
<a href="https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btad561/7271182" class="external-link">our
paper</a>. They implement Phylofit in Rstan, in a similar manner to the
birth-death MCMC. First, we compile the Stan code:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stan.code</span> <span class="op">&lt;-</span> <span class="st">"functions{</span></span>
<span><span class="st">real glogistic_lpdf(real[] t, real S, real tm, real T,real N){</span></span>
<span><span class="st">int n;</span></span>
<span><span class="st">real ll;</span></span>
<span><span class="st">n=size(t);</span></span>
<span><span class="st">ll=0.0;</span></span>
<span><span class="st">for(k in 2:n){</span></span>
<span><span class="st">ll=ll+log(1+exp(S*(t[k-1]-T+tm)));</span></span>
<span><span class="st">ll=ll-(1/(S*N))*choose(k,2)*exp(-S*(T-tm-t[k]))*(exp(S*(t[k-1]-t[k]))-1);</span></span>
<span><span class="st">ll=ll+choose(k,2)*(t[k]-t[k-1])/N+log(choose(k,2))-log(N);</span></span>
<span><span class="st">}</span></span>
<span><span class="st">return(ll);</span></span>
<span><span class="st">}</span></span>
<span><span class="st">}</span></span>
<span><span class="st">data{</span></span>
<span><span class="st">int N;      //number of tips in the tree</span></span>
<span><span class="st">real t[N];  //timing of coalescences (with an added, unused zero)</span></span>
<span><span class="st">real T;</span></span>
<span><span class="st">real maxT;</span></span>
<span><span class="st">real minT;</span></span>
<span><span class="st">real maxLN;</span></span>
<span><span class="st">real minLN;</span></span>
<span><span class="st">}</span></span>
<span><span class="st">parameters {</span></span>
<span><span class="st">real&lt;lower=0.0001,upper=4&gt; S; //instantaneous growth rate-&gt; exp(S)-1 per Year.</span></span>
<span><span class="st">real &lt;lower=minT,upper=maxT&gt; tm; //midpoint</span></span>
<span><span class="st">real&lt;lower=minLN,upper=maxLN&gt; LN; //log (base 10) pop size (i.e. carrying capacity)</span></span>
<span><span class="st">}</span></span>
<span><span class="st">model {</span></span>
<span><span class="st">S ~ uniform(0.001,4);</span></span>
<span><span class="st">tm ~ uniform(minT,maxT);</span></span>
<span><span class="st">LN ~ uniform(minLN,maxLN);</span></span>
<span><span class="st">t ~ glogistic(S,tm,T,pow(10,LN));</span></span>
<span><span class="st">}</span></span>
<span><span class="st">"</span></span>
<span><span class="va">PHYLOFIT_STAN_MODEL</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mc-stan.org/rstan/reference/stan_model.html" class="external-link">stan_model</a></span><span class="op">(</span>model_code <span class="op">=</span> <span class="va">stan.code</span>, model_name <span class="op">=</span> <span class="st">"Phylofit"</span><span class="op">)</span></span></code></pre></div>
<p>Now that the Stan code is compiled, it’s ready to be run. We’ll wrap
it in a function similar to our other functions for estimating the
growth rate. We’ve made some cosmetic changes, but nothing that will
affect the Phylofit (no ACF) output for the estimated growth rate:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Note that this will only work for binary trees</span></span>
<span><span class="va">phylofit</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ultratree</span>, <span class="va">nChain</span> <span class="op">=</span> <span class="fl">3</span>, <span class="va">nCores</span> <span class="op">=</span> <span class="fl">1</span>, <span class="va">minLN</span> <span class="op">=</span> <span class="fl">4</span>, <span class="va">maxLN</span> <span class="op">=</span> <span class="fl">7</span>,</span>
<span>                     <span class="va">chainLength</span> <span class="op">=</span> <span class="fl">20000</span>, <span class="va">alpha</span> <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">ptm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Get data from the tree</span></span>
<span>  <span class="va">coal_times</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/branching.times.html" class="external-link">branching.times</a></span><span class="op">(</span><span class="va">ultratree</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span>  <span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"N"</span> <span class="op">=</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/summary.phylo.html" class="external-link">Ntip</a></span><span class="op">(</span><span class="va">ultratree</span><span class="op">)</span>,</span>
<span>    <span class="st">"t"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">coal_times</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="st">"T"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">coal_times</span><span class="op">)</span>,</span>
<span>    <span class="st">"maxT"</span> <span class="op">=</span> <span class="fl">2</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">coal_times</span><span class="op">)</span>,</span>
<span>    <span class="st">"minLN"</span> <span class="op">=</span> <span class="va">minLN</span>, <span class="st">"maxLN"</span> <span class="op">=</span> <span class="va">maxLN</span></span>
<span>  <span class="op">)</span></span>
<span>  <span class="va">dat</span><span class="op">$</span><span class="va">minT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="cn">T</span> <span class="op">-</span> <span class="va">dat</span><span class="op">$</span><span class="va">t</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Run the Stan code</span></span>
<span>  <span class="va">stanr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mc-stan.org/rstan/reference/stanmodel-method-sampling.html" class="external-link">sampling</a></span><span class="op">(</span><span class="va">PHYLOFIT_STAN_MODEL</span>,</span>
<span>    data <span class="op">=</span> <span class="va">dat</span>, iter <span class="op">=</span> <span class="va">chainLength</span>, control <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>adapt_delta <span class="op">=</span> <span class="fl">0.95</span><span class="op">)</span>,</span>
<span>    chains <span class="op">=</span> <span class="va">nChain</span>, cores <span class="op">=</span> <span class="va">nCores</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Return the results</span></span>
<span>  <span class="va">ptile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">alpha</span> <span class="op">/</span> <span class="fl">2</span>, <span class="fl">.5</span>, <span class="fl">1</span> <span class="op">-</span> <span class="va">alpha</span> <span class="op">/</span> <span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="va">posterior</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="fu">rstan</span><span class="fu">::</span><span class="fu"><a href="https://mc-stan.org/rstan/reference/stanfit-method-extract.html" class="external-link">extract</a></span><span class="op">(</span><span class="va">stanr</span><span class="op">)</span><span class="op">$</span><span class="va">S</span>, <span class="va">ptile</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    <span class="st">"lowerBound"</span> <span class="op">=</span> <span class="va">posterior</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"estimate"</span> <span class="op">=</span> <span class="va">posterior</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>    <span class="st">"upperBound"</span> <span class="op">=</span> <span class="va">posterior</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>, <span class="st">"runtime_s"</span> <span class="op">=</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">ptm</span><span class="op">)</span><span class="op">[</span><span class="st">"elapsed"</span><span class="op">]</span>,</span>
<span>    <span class="st">"n"</span> <span class="op">=</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/summary.phylo.html" class="external-link">Ntip</a></span><span class="op">(</span><span class="va">ultratree</span><span class="op">)</span>, <span class="st">"alpha"</span> <span class="op">=</span> <span class="va">alpha</span>, <span class="st">"method"</span> <span class="op">=</span> <span class="st">"Phylofit no ACF"</span>,</span>
<span>    <span class="st">"nChains"</span> <span class="op">=</span> <span class="va">nChain</span>, <span class="st">"nCores"</span> <span class="op">=</span> <span class="va">nCores</span>, <span class="st">"chainLength"</span> <span class="op">=</span> <span class="va">chainLength</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="varying-parameters">Varying parameters<a class="anchor" aria-label="anchor" href="#varying-parameters"></a>
</h2>
<p>So far, we’ve looked at the same parameters. Now, we’ll see what
happens when we change some important parameters about the birth-death
process.</p>
<div class="section level3">
<h3 id="varying-n">Varying n<a class="anchor" aria-label="anchor" href="#varying-n"></a>
</h3>
<p>One of the most interesting results to explore further is how the
accuracy of the estimates depends on the number of samples. For this,
we’ll look at a few different <code>n</code> parameter values. We can
run <code><a href="../reference/simUltra.html">simUltra()</a></code> with a vector input for <code>n</code>.
First, we’ll generate the vector, which we’ll call <code>n_vec</code>.
Keeping with the paper, we’ll produce 500 trees for each value of n:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">30</span>, <span class="fl">50</span>, <span class="fl">100</span>, <span class="fl">200</span>, <span class="fl">500</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span></code></pre></div>
<p>We keep the growth rate <code>a - b</code> equal to 0.5, but let the
actual birth and death rates vary.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">a_vec</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">n_vec</span><span class="op">)</span>, min <span class="op">=</span> <span class="fl">0.5</span>, max <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span>
<span><span class="va">b_vec</span> <span class="op">&lt;-</span> <span class="va">a_vec</span> <span class="op">-</span> <span class="fl">0.5</span></span></code></pre></div>
<p>In total, our <code>n_vec</code> is of length 3000. So we’ll want to
make sure <code>a_vec</code> and <code>b_vec</code> are also of length
3000. And we’ll have to set <code>nTrees = 3000</code>. If we generate
multiple trees, we must either set a single value for the parameters, or
a vector of length <code>nTrees</code>. This applies to both
<code><a href="../reference/simUltra.html">simUltra()</a></code> and <code><a href="../reference/simMut.html">simMut()</a></code>. Let’s generate some
more trees. Note: this will take a few minutes, but can be parallelized
by setting nCores &gt; 1.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vary_n_trees</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simUltra.html">simUltra</a></span><span class="op">(</span></span>
<span>  a <span class="op">=</span> <span class="va">a_vec</span>, b <span class="op">=</span> <span class="va">b_vec</span>, cloneAge <span class="op">=</span> <span class="fl">40</span>, n <span class="op">=</span> <span class="va">n_vec</span>,</span>
<span>  nTrees <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">n_vec</span><span class="op">)</span>, nCores <span class="op">=</span> <span class="fl">6</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Now, let’s apply our estimates and combine the data.frames:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Apply our estimates for ultrametric trees</span></span>
<span><span class="va">vary_n_maxLike</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="../reference/maxLikelihood.html">maxLikelihood</a></span><span class="op">(</span><span class="va">vary_n_trees</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">vary_n_lengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/internalLengths.html">internalLengths</a></span><span class="op">(</span><span class="va">vary_n_trees</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combine the estimates</span></span>
<span><span class="va">vary_n_analytical</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">vary_n_maxLike</span>, <span class="va">vary_n_lengths</span><span class="op">)</span></span></code></pre></div>
<p>Let’s make a plot for each of these, showing the performance. We’ll
use <code><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">ggplot2::facet_wrap()</a></code> in order to show the same plot at
various <code>n</code> values.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">vary_n_analytical</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">estimate</span>, color <span class="op">=</span> <span class="va">method</span><span class="op">)</span>, linewidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, axis.ticks.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Net growth rate estimate (r)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Density"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span></span>
<span>    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Internal lengths"</span>, <span class="st">"Max. likelihood"</span><span class="op">)</span>,</span>
<span>    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="va">colorPal</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"n = "</span>, <span class="va">n</span><span class="op">)</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"n = "</span>, <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">n_vec</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    ncol <span class="op">=</span> <span class="fl">1</span>, strip.position <span class="op">=</span> <span class="st">"bottom"</span>, scales <span class="op">=</span> <span class="st">"free"</span>, dir <span class="op">=</span> <span class="st">"v"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 2 rows containing non-finite values (`stat_density()`).</span></span></code></pre></div>
<p><img src="reproduce-simResults_files/figure-html/plot_vary_n-1.png" width="100%"></p>
<p>As you can see, both estimates are highly dependent on the number of
samples. If you’d like to see the same density values on the y-axis for
all <code>n</code> values, set <code>scales = "fixed"</code> in the
<code><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap()</a></code> function.</p>
<div class="section level4">
<h4 id="markov-chain-monte-carlo-mcmc-application">Markov Chain Monte Carlo (MCMC) application<a class="anchor" aria-label="anchor" href="#markov-chain-monte-carlo-mcmc-application"></a>
</h4>
<p>Note: computationally intensive! Recommended to run in high
performance computing environment.</p>
<p>In our paper, we show two alternatives, each with their own set of
advantages and drawbacks. The drawback for both MCMC approaches is that
the computational expense is greatly increased compared to our
<code><a href="../reference/internalLengths.html">internalLengths()</a></code> and <code><a href="../reference/maxLikelihood.html">maxLikelihood()</a></code>
estimates. Phylofit has the additional drawback of overly confident
confidence intervals, where the 95% Highest Posterior Density (HPD)
doesn’t capture the true value 95% of the time. This is because the
deterministic approximation of the population size trajectory in
Phylofit doesn’t match the stochastic underlying birth-death process
which generates the simulated trees. The <code><a href="../reference/birthDeathMCMC.html">birthDeathMCMC()</a></code>
approach doesn’t suffer from this drawback.</p>
<p>With both MCMC approaches, the added computational expense comes with
a notable benefit. These methods do not require that the birth-death
process is supercritical, and will work well even for small growth
rates. We show this in the section <a href="#varying-r-andor-cloneage">Varying r (and/or cloneAge)</a>. While
Phylofit does have an additional drawback, it also has increased
flexibility. Phylofit will work well in estimating the growth rate even
if the coalescence times are characteristic of a logistic clonal growth
trajectory. In most cases, the number of samples will be small enough
and the effective population size large enough that the coalescence
times date back to the exponential phase, before carrying capacity and
logistic growth (or other saturating growth) affect the coalescence
times, in which case, constant birth-death approximations
(<code><a href="../reference/birthDeathMCMC.html">birthDeathMCMC()</a></code>, <code><a href="../reference/maxLikelihood.html">maxLikelihood()</a></code>, and
<code><a href="../reference/internalLengths.html">internalLengths()</a></code>) work well.</p>
<p>This is how we would run our MCMC’s. We include the birth-death MCMC
as a part of the <code>cloneRate</code> package, so it’s a
straightforward one-liner:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># NOTE: depending on the number of cores, this will take hours to days</span></span>
<span><span class="va">vary_n_bdMCMC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/birthDeathMCMC.html">birthDeathMCMC</a></span><span class="op">(</span><span class="va">vary_n_trees</span>,</span>
<span>  nCores <span class="op">=</span> <span class="fl">3</span>, nChains <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  chainLength <span class="op">=</span> <span class="fl">15000</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>And for Phylofit, we use the <code>phylofit()</code> function we
created at the start, which is a copy of the function from Emily
Mitchell’s github and was created by Nick Williams and used in Williams
et al. and Mitchell et al.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># NOTE: depending on the number of cores, this will take hours to days</span></span>
<span><span class="va">vary_n_phylofit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span></span>
<span>  <span class="va">rbind</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/parallel/mclapply.html" class="external-link">mclapply</a></span><span class="op">(</span><span class="va">vary_n_trees</span>, <span class="va">phylofit</span>,</span>
<span>    mc.cores <span class="op">=</span> <span class="fl">5</span>, mc.silent <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Now we can make a list for each set of 500 trees, split by method and
number of samples, <code>n</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Make a list with each list element containing a data.frame of results for a</span></span>
<span><span class="co"># given n and method</span></span>
<span><span class="va">vary_n_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">vary_n_lengths</span>, <span class="va">vary_n_lengths</span><span class="op">$</span><span class="va">n</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">vary_n_maxLike</span>, <span class="va">vary_n_maxLike</span><span class="op">$</span><span class="va">n</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">vary_n_bdMCMC</span>, <span class="va">vary_n_bdMCMC</span><span class="op">$</span><span class="va">n</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">vary_n_phylofit</span>, <span class="va">vary_n_phylofit</span><span class="op">$</span><span class="va">n</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  recursive <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="summary-statistics">Summary statistics<a class="anchor" aria-label="anchor" href="#summary-statistics"></a>
</h4>
<p>We can write a basic function to give us all the summary statistics
we want from the estimates. Make sure that each input data.frame
contains all the same number of tips, true growth rate, method, etc. so
that it should be combined.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">get_summary_stats</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">df</span>, <span class="va">growthRate</span>, <span class="va">cloneAge</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">meanEstimate</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">estimate</span><span class="op">)</span></span>
<span>  <span class="va">sdEstimate</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/sd.html" class="external-link">sd</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">estimate</span><span class="op">)</span></span>
<span>  <span class="va">medianEstimate</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">estimate</span><span class="op">)</span></span>
<span>  <span class="va">normalizedRMSE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">(</span><span class="op">(</span><span class="va">growthRate</span> <span class="op">-</span> <span class="va">df</span><span class="op">$</span><span class="va">estimate</span><span class="op">)</span> <span class="op">/</span> <span class="va">growthRate</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">coverageProb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">lowerBound</span> <span class="op">&lt;</span> <span class="va">growthRate</span> <span class="op">&amp;</span> <span class="va">df</span><span class="op">$</span><span class="va">upperBound</span> <span class="op">&gt;</span> <span class="va">growthRate</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span>
<span>  <span class="va">runtimeMean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">runtime_s</span><span class="op">)</span></span>
<span>  <span class="va">runtimeSD</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/sd.html" class="external-link">sd</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">runtime_s</span><span class="op">)</span></span>
<span>  <span class="va">method</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">method</span><span class="op">)</span></span>
<span>  <span class="va">nCores</span> <span class="op">&lt;-</span> <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">nCores</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fl">1</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">df</span><span class="op">$</span><span class="va">nCores</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">nChains</span> <span class="op">&lt;-</span> <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">nChains</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="cn">NA</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">df</span><span class="op">$</span><span class="va">nChains</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">nTips</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">n</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Combine everything into a data.frame to return</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    <span class="st">"Method"</span> <span class="op">=</span> <span class="va">method</span>, <span class="st">"Number of samples, n"</span> <span class="op">=</span> <span class="va">nTips</span>,</span>
<span>    <span class="st">"Net growth rate, r"</span> <span class="op">=</span> <span class="va">growthRate</span>, <span class="st">"Clone age, T"</span> <span class="op">=</span> <span class="va">cloneAge</span>,</span>
<span>    <span class="st">"Number of simulated trees"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span>,</span>
<span>    <span class="st">"Mean estimate"</span> <span class="op">=</span> <span class="va">meanEstimate</span>, <span class="st">"SD estimate"</span> <span class="op">=</span> <span class="va">sdEstimate</span>,</span>
<span>    <span class="st">"Median Estimate"</span> <span class="op">=</span> <span class="va">medianEstimate</span>,</span>
<span>    <span class="st">"Normalized RMSE"</span> <span class="op">=</span> <span class="va">normalizedRMSE</span>,</span>
<span>    <span class="st">"95% Coverage probability"</span> <span class="op">=</span> <span class="va">coverageProb</span>, <span class="co"># Assuming 95% CI</span></span>
<span>    <span class="st">"Runtime (s)"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">runtimeMean</span>, <span class="fl">3</span><span class="op">)</span>, <span class="st">" +/- "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">runtimeSD</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="st">"Number of cores"</span> <span class="op">=</span> <span class="va">nCores</span>, <span class="st">"Number of chains"</span> <span class="op">=</span> <span class="va">nChains</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>We can apply our summary stats function to the list of results
contained in <code>vary_n_results</code> to get a summary data.frame
with statistics for each value of <code>n</code> and each method.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">summary_n.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">vary_n_results</span>, <span class="va">get_summary_stats</span>,</span>
<span>  growthRate <span class="op">=</span> <span class="fl">0.5</span>, cloneAge <span class="op">=</span> <span class="fl">40</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The information in <code>summary_n.df</code> is available in
Supplementary table 6 of our paper, and is also available as “.csv”
file. Note that the stochastic nature of the tree generation and MCMCs
will lead to slight differences in the results each time the analysis is
performed.</p>
</div>
</div>
<div class="section level3">
<h3 id="varying-r-andor-cloneage">Varying r (and/or cloneAge)<a class="anchor" aria-label="anchor" href="#varying-r-andor-cloneage"></a>
</h3>
<p>Here, we’ll recreate the analysis in Figures 3 and 4 of <a href="https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btad561/7271182" class="external-link">our
paper</a>, noting that our methods struggle with small growth rates.</p>
<div class="section level4">
<h4 id="varying-r-or-cloneage">Varying r or cloneAge?<a class="anchor" aria-label="anchor" href="#varying-r-or-cloneage"></a>
</h4>
<p>First, we have to address something that might otherwise lead to
confusion; varying <code>r</code> is the same as varying
<code>cloneAge</code>. If I simulate a population with a net growth rate
of 1 per year for 20 years it should look the same as a growth rate of
0.5 per year for 40 years. We’ll show this, but first consider the fact
that the units are meaningless, so long as they’re consistent. So a
growth rate of 1/12 and a clone age of 20*12 is the same as a growth
rate of 1 and 20 if the units change from years to months.</p>
<p>This all might be a bit confusing, so let’s plot it. First, we’ll
simulate a tree with a growth rate of 2 for 30 years and compare it to a
tree with a growth rate of 0.5 and the same 30 years. These trees should
look different. Finally, we’ll add a tree that has a growth rate of 0.5,
but run it for 120 years. If what I said above is true, we should see
that the last tree looks like the first tree, because 0.5 for 120 should
be the same as 2 for 30. Let’s find out:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First tree, r = a - b = 2</span></span>
<span><span class="va">tree1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simUltra.html">simUltra</a></span><span class="op">(</span>a <span class="op">=</span> <span class="fl">2.5</span>, b <span class="op">=</span> <span class="fl">.5</span>, cloneAge <span class="op">=</span> <span class="fl">30</span>, n <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Second tree, with r = a - b = 0.5</span></span>
<span><span class="va">tree2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simUltra.html">simUltra</a></span><span class="op">(</span>a <span class="op">=</span> <span class="fl">1</span>, b <span class="op">=</span> <span class="fl">.5</span>, cloneAge <span class="op">=</span> <span class="fl">30</span>, n <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Third tree, with r = 0.5 but cloneAge = 120</span></span>
<span><span class="va">tree3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simUltra.html">simUltra</a></span><span class="op">(</span>a <span class="op">=</span> <span class="fl">1</span>, b <span class="op">=</span> <span class="fl">.5</span>, cloneAge <span class="op">=</span> <span class="fl">120</span>, n <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p>Now, let’s plot using <code><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par()</a></code> to show all three in one
plot</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">graphics</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/plot.phylo.html" class="external-link">plot.phylo</a></span><span class="op">(</span><span class="va">tree1</span>, direction <span class="op">=</span> <span class="st">"downwards"</span>, show.tip.label <span class="op">=</span> <span class="cn">F</span>, main <span class="op">=</span> <span class="st">"r = 2, cloneAge = 30"</span><span class="op">)</span></span>
<span><span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/plot.phylo.html" class="external-link">plot.phylo</a></span><span class="op">(</span><span class="va">tree2</span>, direction <span class="op">=</span> <span class="st">"downwards"</span>, show.tip.label <span class="op">=</span> <span class="cn">F</span>, main <span class="op">=</span> <span class="st">"r = 0.5, cloneAge = 30"</span><span class="op">)</span></span>
<span><span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/plot.phylo.html" class="external-link">plot.phylo</a></span><span class="op">(</span><span class="va">tree3</span>, direction <span class="op">=</span> <span class="st">"downwards"</span>, show.tip.label <span class="op">=</span> <span class="cn">F</span>, main <span class="op">=</span> <span class="st">"r = 0.5, cloneAge = 120"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reproduce-simResults_files/figure-html/plot_vary_r-1.png" width="100%"></p>
<p>While the trees on the left and the right aren’t identical due to the
stochastic nature of the birth-death process, they are similar, and
certainly different from the tree in the middle. We note that there are
slight differences in the distribution of coalescence times due to
different birth and death rates (i.e. different extinction
probabilities), but these differences are small enough to be
negligible.</p>
</div>
<div class="section level4">
<h4 id="performance-across-r-values">Performance across r values<a class="anchor" aria-label="anchor" href="#performance-across-r-values"></a>
</h4>
<p>Now we can arbitrarily choose whether to vary <code>r</code> or
<code>cloneAge</code>. For consistency, we’ll vary <code>r</code>, as we
do in Figures 3 and 4 of <a href="https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btad561/7271182" class="external-link">our
paper</a>. Here, we have two analyses:</p>
<ol style="list-style-type: decimal">
<li>Simulate at fixed <code>r</code> values, showing which are good and
which are problematic</li>
<li>Simulate at random <code>r</code> values and then try to decipher
which estimates are good and which aren’t</li>
</ol>
<div class="section level5">
<h5 id="set-r-fig--3">Set r (Fig. 3)<a class="anchor" aria-label="anchor" href="#set-r-fig--3"></a>
</h5>
<p>This is essentially the same as we did in <a href="#varying-n">Varying n</a>, but with <code>r</code> instead of
<code>n</code>, so we can run through this quickly. First, generate the
trees, which all have <code>n=100</code> tips and an age of 40 years,
but vary growth rate, from 0.15 to 1.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">r_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.15</span>, <span class="fl">0.2</span>, <span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="fl">0.75</span>, <span class="fl">1</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">a_vec</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">r_vec</span><span class="op">)</span>, min <span class="op">=</span> <span class="va">r_vec</span>, max <span class="op">=</span> <span class="va">r_vec</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">b_vec</span> <span class="op">&lt;-</span> <span class="va">a_vec</span> <span class="op">-</span> <span class="va">r_vec</span></span>
<span></span>
<span><span class="va">vary_r_trees</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simUltra.html">simUltra</a></span><span class="op">(</span></span>
<span>  a <span class="op">=</span> <span class="va">a_vec</span>, b <span class="op">=</span> <span class="va">b_vec</span>, cloneAge <span class="op">=</span> <span class="fl">40</span>, n <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  nTrees <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">a_vec</span><span class="op">)</span>, nCores <span class="op">=</span> <span class="fl">6</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Again, we keep the growth rate <code>a - b</code> equal to
<code>r</code>, but let the actual birth and death rates vary. Now we
apply our methods for estimating the growth rate:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Apply our fast estimates (computationally cheap)</span></span>
<span><span class="va">vary_r_lengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="../reference/internalLengths.html">internalLengths</a></span><span class="op">(</span><span class="va">vary_r_trees</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">vary_r_maxLike</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="../reference/maxLikelihood.html">maxLikelihood</a></span><span class="op">(</span><span class="va">vary_r_trees</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>In the above chunk we suppress warnings related to the diagnostic
which tells us whether our methods are applicable. We discuss this
further in [Random r].</p>
<p>NOTE: As above, running the MCMC methods many times is
computationally expensive and should be done in a high performance
computing environment:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># NOTE: depending on the number of cores, these will take hours to days</span></span>
<span><span class="va">vary_r_bdMCMC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/birthDeathMCMC.html">birthDeathMCMC</a></span><span class="op">(</span><span class="va">vary_r_trees</span>,</span>
<span>  nCores <span class="op">=</span> <span class="fl">3</span>, nChains <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  chainLength <span class="op">=</span> <span class="fl">15000</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">vary_r_phylofit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span></span>
<span>  <span class="va">rbind</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/parallel/mclapply.html" class="external-link">mclapply</a></span><span class="op">(</span><span class="va">vary_r_trees</span>, <span class="va">phylofit</span>,</span>
<span>    mc.cores <span class="op">=</span> <span class="fl">5</span>, mc.silent <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make a list with each list element containing a data.frame of results for a</span></span>
<span><span class="co"># given r and method</span></span>
<span><span class="va">vary_r_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">vary_r_lengths</span>, <span class="va">r_vec</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">vary_r_maxLike</span>, <span class="va">r_vec</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">vary_r_bdMCMC</span>, <span class="va">r_vec</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">vary_r_phylofit</span>, <span class="va">r_vec</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  recursive <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">summary_r.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="va">get_summary_stats</span>, <span class="va">vary_r_results</span>,</span>
<span>  growthRate <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">vary_r_results</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  cloneAge <span class="op">=</span> <span class="fl">40</span>, SIMPLIFY <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="random-r-fig--4">Random r (Fig. 4)<a class="anchor" aria-label="anchor" href="#random-r-fig--4"></a>
</h5>
<p>In order to be able to apply our estimates in real-world cases where
the growth rate is not known, we need a diagnostic to tell us whether
our methods are applicable. This diagnostic should tell us if the tree
is supercritical enough for our approximations to hold. The ratio of
external to internal lengths is a measure of how “supercritical” or
star-shaped the birth-death process is. If the internal lengths are at
the top of the tree, we consider it star-shaped, and it’s a good
candidate for our methods, which operate under the assumption of a
supercritical process. This means that the growth rate is high enough,
and enough time has passed, for the coalescence events to occur well
before the sampling time. In practice, using these simulations, we show
that an external to internal lengths ratio greater than 3 is good enough
to apply our methods. Let’s apply our methods and see how they look
before and after applying this ratio cutoff. Note that this is what
we’re doing in Figure 4 of <a href="https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btad561/7271182" class="external-link">our
paper</a>.</p>
<p>Let’s simulate 10,000 trees with 50 samples each and various growth
rates, ranging from 0.1 to 1 per year, run for 40 years. Then, we should
be able to dial in the ratio of external to internal lengths that is
large enough to work well with our methods.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Uniform ditribution of r used to generate b_vec</span></span>
<span><span class="va">r_vec</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10000</span>, min <span class="op">=</span> <span class="fl">0.1</span>, max <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">a_vec</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10000</span>, min <span class="op">=</span> <span class="fl">1</span>, max <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">b_vec</span> <span class="op">&lt;-</span> <span class="va">a_vec</span> <span class="op">-</span> <span class="va">r_vec</span></span>
<span></span>
<span><span class="co"># Input to simUltra()</span></span>
<span><span class="va">random_r_trees</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simUltra.html">simUltra</a></span><span class="op">(</span></span>
<span>  a <span class="op">=</span> <span class="va">a_vec</span>, b <span class="op">=</span> <span class="va">b_vec</span>, cloneAge <span class="op">=</span> <span class="fl">40</span>, n <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  nTrees <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">a_vec</span><span class="op">)</span>, nCores <span class="op">=</span> <span class="fl">6</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Apply our <code><a href="../reference/internalLengths.html">internalLengths()</a></code> and
<code><a href="../reference/maxLikelihood.html">maxLikelihood()</a></code> functions as usual, except that we want to
suppress the warnings. These warnings will tell us that the external to
internal lengths ratio is not high enough for some of the trees, which
is exactly what we’re testing here.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">random_r_maxLike</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="../reference/maxLikelihood.html">maxLikelihood</a></span><span class="op">(</span><span class="va">random_r_trees</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">random_r_lengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="../reference/internalLengths.html">internalLengths</a></span><span class="op">(</span><span class="va">random_r_trees</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Because we have different <code>r</code> values, we can’t do a
density plot as before. We’ll have to show a density plot with the x
axis representing percent error. In this case, good estimates will
consolidate around 0. First, we’ll calculate error using the
<code>estimate</code> column and <code>r_vec</code> which has our ground
truth values.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate error in lengths, adding columns</span></span>
<span><span class="va">random_r_lengths</span><span class="op">$</span><span class="va">ground_truth</span> <span class="op">&lt;-</span> <span class="va">r_vec</span></span>
<span><span class="va">random_r_lengths</span><span class="op">$</span><span class="va">percent_error</span> <span class="op">&lt;-</span> <span class="fl">100</span> <span class="op">*</span> <span class="op">(</span><span class="va">random_r_lengths</span><span class="op">$</span><span class="va">estimate</span> <span class="op">-</span> <span class="va">random_r_lengths</span><span class="op">$</span><span class="va">ground_truth</span><span class="op">)</span> <span class="op">/</span> <span class="va">random_r_lengths</span><span class="op">$</span><span class="va">ground_truth</span></span>
<span></span>
<span><span class="co"># Do the same for maxLike</span></span>
<span><span class="va">random_r_maxLike</span><span class="op">$</span><span class="va">ground_truth</span> <span class="op">&lt;-</span> <span class="va">r_vec</span></span>
<span><span class="va">random_r_maxLike</span><span class="op">$</span><span class="va">percent_error</span> <span class="op">&lt;-</span> <span class="fl">100</span> <span class="op">*</span> <span class="op">(</span><span class="va">random_r_maxLike</span><span class="op">$</span><span class="va">estimate</span> <span class="op">-</span> <span class="va">random_r_maxLike</span><span class="op">$</span><span class="va">ground_truth</span><span class="op">)</span> <span class="op">/</span> <span class="va">random_r_maxLike</span><span class="op">$</span><span class="va">ground_truth</span></span>
<span></span>
<span><span class="co"># Combine for ggplot formatting</span></span>
<span><span class="va">results_random_r</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">random_r_lengths</span>, <span class="va">random_r_maxLike</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot, adding a vertical line at 0 because that's the error free estimate</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">results_random_r</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">percent_error</span>, color <span class="op">=</span> <span class="va">method</span><span class="op">)</span>, linewidth <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>      x <span class="op">=</span> <span class="va">percent_error</span>, color <span class="op">=</span> <span class="va">method</span>,</span>
<span>      y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">results_random_r</span><span class="op">)</span>, min <span class="op">=</span> <span class="op">-</span><span class="fl">.01</span>, max <span class="op">=</span> <span class="op">-</span><span class="fl">.001</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">.1</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="op">-</span><span class="fl">100</span>, <span class="fl">2000</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">100</span>, <span class="fl">150</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, axis.ticks.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Estimate percent error"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Density"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Internal lengths"</span>, <span class="st">"Max. likelihood"</span><span class="op">)</span>, values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"#009E73"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reproduce-simResults_files/figure-html/coverage_vary_r-1.png" width="100%"></p>
<p>Not bad, but we have some rogue estimates with &gt; 150% error! And
we have many with &gt; 50% error.</p>
<p>The external to internal lengths ratio should give a good estimate of
how supercritical a tree is. But we need to figure out what ratio is
good enough for our assumptions to be valid and our methods to work. For
each method, let’s split the results by the ratio of external to
internal lengths, and see how accurate the 95% confidence intervals
are:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Split by the integer value of the ratio</span></span>
<span></span>
<span><span class="co"># Make a dataframe to store the coverage probabilities</span></span>
<span><span class="va">covProbs.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  <span class="st">"intervalMin"</span> <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  <span class="st">"intervalMax"</span> <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  <span class="st">"coverageProb"</span> <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  <span class="st">"method"</span> <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  <span class="st">"count"</span> <span class="op">=</span> <span class="cn">NA</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate coverage probability at each interval of ext. to int. lengths ratio</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># Get the subset that has a ratio in the interval i to i+1</span></span>
<span>  <span class="va">lengths.tmp</span> <span class="op">&lt;-</span> <span class="va">random_r_lengths</span><span class="op">[</span><span class="va">random_r_lengths</span><span class="op">$</span><span class="va">extIntRatio</span> <span class="op">&gt;=</span> <span class="va">i</span> <span class="op">&amp;</span></span>
<span>    <span class="va">random_r_lengths</span><span class="op">$</span><span class="va">extIntRatio</span> <span class="op">&lt;</span> <span class="va">i</span> <span class="op">+</span> <span class="fl">1</span>, <span class="op">]</span></span>
<span>  <span class="va">maxLike.tmp</span> <span class="op">&lt;-</span> <span class="va">random_r_maxLike</span><span class="op">[</span><span class="va">random_r_maxLike</span><span class="op">$</span><span class="va">extIntRatio</span> <span class="op">&gt;=</span> <span class="va">i</span> <span class="op">&amp;</span></span>
<span>    <span class="va">random_r_maxLike</span><span class="op">$</span><span class="va">extIntRatio</span> <span class="op">&lt;</span> <span class="va">i</span> <span class="op">+</span> <span class="fl">1</span>, <span class="op">]</span></span>
<span></span>
<span>  <span class="va">coverageLengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">lengths.tmp</span><span class="op">$</span><span class="va">lowerBound</span> <span class="op">&lt;</span> <span class="va">lengths.tmp</span><span class="op">$</span><span class="va">ground_truth</span> <span class="op">&amp;</span></span>
<span>    <span class="va">lengths.tmp</span><span class="op">$</span><span class="va">upperBound</span> <span class="op">&gt;</span> <span class="va">lengths.tmp</span><span class="op">$</span><span class="va">ground_truth</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">lengths.tmp</span><span class="op">)</span></span>
<span>  <span class="va">coverageMaxLike</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">maxLike.tmp</span><span class="op">$</span><span class="va">lowerBound</span> <span class="op">&lt;</span> <span class="va">maxLike.tmp</span><span class="op">$</span><span class="va">ground_truth</span> <span class="op">&amp;</span></span>
<span>    <span class="va">maxLike.tmp</span><span class="op">$</span><span class="va">upperBound</span> <span class="op">&gt;</span> <span class="va">maxLike.tmp</span><span class="op">$</span><span class="va">ground_truth</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">maxLike.tmp</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">covProbs.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">covProbs.df</span>, <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    <span class="st">"intervalMin"</span> <span class="op">=</span> <span class="va">i</span>,</span>
<span>    <span class="st">"intervalMax"</span> <span class="op">=</span> <span class="va">i</span> <span class="op">+</span> <span class="fl">1</span>,</span>
<span>    <span class="st">"coverageProb"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="va">coverageLengths</span>,</span>
<span>      <span class="va">coverageMaxLike</span></span>
<span>    <span class="op">)</span>,</span>
<span>    <span class="st">"method"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lengths"</span>, <span class="st">"maxLike"</span><span class="op">)</span>,</span>
<span>    <span class="st">"count"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">maxLike.tmp</span><span class="op">)</span></span>
<span>  <span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Remove NA row from initialization</span></span>
<span><span class="va">covProbs.df</span> <span class="op">&lt;-</span> <span class="va">covProbs.df</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">covProbs.df</span><span class="op">$</span><span class="va">intervalMin</span><span class="op">)</span>, <span class="op">]</span></span></code></pre></div>
<p>Let’s plot this as we do in Figure 4A of our paper:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">covProbs.df</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html" class="external-link">geom_segment</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">intervalMin</span>, y <span class="op">=</span> <span class="va">coverageProb</span>,</span>
<span>    xend <span class="op">=</span> <span class="va">intervalMax</span>, yend <span class="op">=</span> <span class="va">coverageProb</span>,</span>
<span>    color <span class="op">=</span> <span class="va">method</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">colorPal</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Coverage Probability"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"External to internal lengths ratio"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">3</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0.95</span>, linetype <span class="op">=</span> <span class="st">"dotted"</span>, linewidth <span class="op">=</span> <span class="fl">.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="reproduce-simResults_files/figure-html/plotRatio-1.png" width="100%"></p>
<p>Seeing that the confidence intervals aren’t accurate until the ratio
is at least 3, let’s apply that cutoff ratio of 3 to exclude the trees
which we don’t think are good candidates for our methods. Hopefully,
this will get rid of the very poor estimates.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Cut those trees with ratio &gt; 3</span></span>
<span><span class="va">results_random_r_cut</span> <span class="op">&lt;-</span> <span class="va">results_random_r</span><span class="op">[</span><span class="va">results_random_r</span><span class="op">$</span><span class="va">extIntRatio</span> <span class="op">&gt;=</span> <span class="fl">3</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># How many did we cut</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">results_random_r</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">results_random_r_cut</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1064</span></span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot the estimates after the diagnostic</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">results_random_r_cut</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">percent_error</span>, color <span class="op">=</span> <span class="va">method</span><span class="op">)</span>, linewidth <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>      x <span class="op">=</span> <span class="va">percent_error</span>, color <span class="op">=</span> <span class="va">method</span>,</span>
<span>      y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">results_random_r_cut</span><span class="op">)</span>, min <span class="op">=</span> <span class="op">-</span><span class="fl">.01</span>, max <span class="op">=</span> <span class="op">-</span><span class="fl">.001</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">.1</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="op">-</span><span class="fl">100</span>, <span class="fl">200</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">100</span>, <span class="fl">150</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, axis.ticks.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Estimate percent error"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Density"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Internal lengths"</span>, <span class="st">"Max. likelihood"</span><span class="op">)</span>, values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"#009E73"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reproduce-simResults_files/figure-html/plotCutoff-1.png" width="100%"></p>
<p>I’d say that’s much better. We cut about 1,000 of the 10,000 trees,
eliminating many of the worst estimates. The exact number of trees
removed will change each time the analysis is run. The real advantage
here is that we can look at a tree and calculate this ratio which will
tell us right away if our methods are applicable.</p>
<p>Finally, let’s quantify what we see with Root mean square error and
coverage probability, as we do in Figure 4D. We’ll start with the
lengths method:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the lengths stats before the diagnostic</span></span>
<span><span class="va">normRMSE_before_lengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">(</span><span class="op">(</span><span class="va">random_r_lengths</span><span class="op">$</span><span class="va">estimate</span> <span class="op">-</span> <span class="va">random_r_lengths</span><span class="op">$</span><span class="va">ground_truth</span><span class="op">)</span> <span class="op">/</span></span>
<span>  <span class="va">random_r_lengths</span><span class="op">$</span><span class="va">ground_truth</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">random_r_lengths</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">coverage_before_lengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">random_r_lengths</span><span class="op">$</span><span class="va">upperBound</span> <span class="op">&gt;</span> <span class="va">random_r_lengths</span><span class="op">$</span><span class="va">ground_truth</span> <span class="op">&amp;</span></span>
<span>  <span class="va">random_r_lengths</span><span class="op">$</span><span class="va">lowerBound</span> <span class="op">&lt;</span> <span class="va">random_r_lengths</span><span class="op">$</span><span class="va">ground_truth</span><span class="op">)</span> <span class="op">/</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">random_r_lengths</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Get the lengths stats for those passing the diagnostic</span></span>
<span><span class="va">lengths_passing</span> <span class="op">&lt;-</span> <span class="va">random_r_lengths</span><span class="op">[</span><span class="va">random_r_lengths</span><span class="op">$</span><span class="va">extIntRatio</span> <span class="op">&gt;</span> <span class="fl">3</span>, <span class="op">]</span></span>
<span><span class="va">normRMSE_after_lengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">(</span><span class="op">(</span><span class="va">lengths_passing</span><span class="op">$</span><span class="va">estimate</span> <span class="op">-</span> <span class="va">lengths_passing</span><span class="op">$</span><span class="va">ground_truth</span><span class="op">)</span> <span class="op">/</span></span>
<span>  <span class="va">lengths_passing</span><span class="op">$</span><span class="va">ground_truth</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">lengths_passing</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">coverage_after_lengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">lengths_passing</span><span class="op">$</span><span class="va">upperBound</span> <span class="op">&gt;</span> <span class="va">lengths_passing</span><span class="op">$</span><span class="va">ground_truth</span> <span class="op">&amp;</span></span>
<span>  <span class="va">lengths_passing</span><span class="op">$</span><span class="va">lowerBound</span> <span class="op">&lt;</span> <span class="va">lengths_passing</span><span class="op">$</span><span class="va">ground_truth</span><span class="op">)</span> <span class="op">/</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">lengths_passing</span><span class="op">)</span></span></code></pre></div>
<p>Do the same for the maximum likelihood estimates:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the maxLike stats before the diagnostic</span></span>
<span><span class="va">normRMSE_before_maxLike</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">(</span><span class="op">(</span><span class="va">random_r_maxLike</span><span class="op">$</span><span class="va">estimate</span> <span class="op">-</span> <span class="va">random_r_maxLike</span><span class="op">$</span><span class="va">ground_truth</span><span class="op">)</span> <span class="op">/</span></span>
<span>  <span class="va">random_r_maxLike</span><span class="op">$</span><span class="va">ground_truth</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">random_r_maxLike</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">coverage_before_maxLike</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">random_r_maxLike</span><span class="op">$</span><span class="va">upperBound</span> <span class="op">&gt;</span> <span class="va">random_r_maxLike</span><span class="op">$</span><span class="va">ground_truth</span> <span class="op">&amp;</span></span>
<span>  <span class="va">random_r_maxLike</span><span class="op">$</span><span class="va">lowerBound</span> <span class="op">&lt;</span> <span class="va">random_r_maxLike</span><span class="op">$</span><span class="va">ground_truth</span><span class="op">)</span> <span class="op">/</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">random_r_maxLike</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Get the maxLike stats for those passing the diagnostic</span></span>
<span><span class="va">maxLike_passing</span> <span class="op">&lt;-</span> <span class="va">random_r_maxLike</span><span class="op">[</span><span class="va">random_r_maxLike</span><span class="op">$</span><span class="va">extIntRatio</span> <span class="op">&gt;</span> <span class="fl">3</span>, <span class="op">]</span></span>
<span><span class="va">normRMSE_after_maxLike</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">(</span><span class="op">(</span><span class="va">maxLike_passing</span><span class="op">$</span><span class="va">estimate</span> <span class="op">-</span> <span class="va">maxLike_passing</span><span class="op">$</span><span class="va">ground_truth</span><span class="op">)</span> <span class="op">/</span></span>
<span>  <span class="va">maxLike_passing</span><span class="op">$</span><span class="va">ground_truth</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">maxLike_passing</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">coverage_after_maxLike</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">maxLike_passing</span><span class="op">$</span><span class="va">upperBound</span> <span class="op">&gt;</span> <span class="va">maxLike_passing</span><span class="op">$</span><span class="va">ground_truth</span> <span class="op">&amp;</span></span>
<span>  <span class="va">maxLike_passing</span><span class="op">$</span><span class="va">lowerBound</span> <span class="op">&lt;</span> <span class="va">maxLike_passing</span><span class="op">$</span><span class="va">ground_truth</span><span class="op">)</span> <span class="op">/</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">maxLike_passing</span><span class="op">)</span></span></code></pre></div>
<p>And then we can construct a table like in Figure 4D:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  <span class="st">"Value"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"RMSE before"</span>, <span class="st">"RMSE after"</span>, <span class="st">"Coverage probability before"</span>,</span>
<span>    <span class="st">"Coverage probability after"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  <span class="st">"Lengths"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="va">normRMSE_before_lengths</span>, <span class="va">normRMSE_after_lengths</span>,</span>
<span>    <span class="va">coverage_before_lengths</span>, <span class="va">coverage_after_lengths</span></span>
<span>  <span class="op">)</span>,</span>
<span>  <span class="st">"Max likelihood"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="va">normRMSE_before_maxLike</span>, <span class="va">normRMSE_after_maxLike</span>,</span>
<span>    <span class="va">coverage_before_maxLike</span>, <span class="va">coverage_after_maxLike</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;                         Value   Lengths Max.likelihood</span></span>
<span><span class="co">#&gt; 1                 RMSE before 0.3920462      0.4570803</span></span>
<span><span class="co">#&gt; 2                  RMSE after 0.1629709      0.1587368</span></span>
<span><span class="co">#&gt; 3 Coverage probability before 0.9021000      0.8696000</span></span>
<span><span class="co">#&gt; 4  Coverage probability after 0.9507610      0.9301701</span></span></code></pre></div>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>The method for simulating the trees is a direct result of the work of
Amaury Lambert in the following paper. The mathematical methods for
estimating growth rates build in large part from the same work, linked
below:</p>
<ul>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/29704514/" class="external-link">Lambert, A.
2018</a></li>
</ul>
<p>Phylofit is used for comparison and is a direct result of previous
work:</p>
<ul>
<li><a href="https://www.nature.com/articles/s41586-021-04312-6" class="external-link">Williams et
al. 2022</a></li>
<li><a href="https://www.nature.com/articles/s41586-022-04786-y" class="external-link">Mitchell et
al. 2022</a></li>
</ul>
<p>The likelihood for the birth-death MCMC was given by Stadler in Eq. 5
of:</p>
<ul>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/19631666/" class="external-link">Stadler, T.
2009</a></li>
</ul>
<p>And here’s a final link to <a href="https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btad561/7271182" class="external-link">our
paper</a> for more of the details of the methods and data analysis.</p>
<p>There aren’t too many colors in this article, but we tried to use
colorblind friendly colors, specifically pulling colors from a palette
designed by <a href="https://www.nature.com/articles/nmeth.1618" class="external-link">Bang
Wong</a> and available <a href="https://davidmathlogic.com/colorblind/#%23000000-%23E69F00-%2356B4E9-%23009E73-%23F0E442-%230072B2-%23D55E00-%23CC79A7" class="external-link">here.</a></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Brian Johnson, Yubo Shuai, Jason Schweinsberg, Kit Curtius.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
